{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "collapsed_sections": [
        "vcr5CjibHTgF",
        "XUZS4GUkHIEG",
        "-z4GQiMfU1wf",
        "g8jcfQ14mA41"
      ]
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# What causes antibiotics resistance?\n",
        "*Мельникова Дана, Чеченина Анна*\n",
        "\n",
        "Ниже перечисленны все команды и результаты, которые мы использовали в своей работе. Все комманды были продублированы локально и финальные результаты, приведенные в отчете, соответствуют прогону на **Ubuntu 18.04.6 LTS bionic**."
      ],
      "metadata": {
        "id": "bRsCeqvCQIig"
      }
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "vcr5CjibHTgF"
      },
      "source": [
        "## Установка FastQC"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "62l1Q2ymbENA",
        "outputId": "8c30747e-859e-4826-82cd-446faba12e49"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "openjdk version \"11.0.16\" 2022-07-19\n",
            "OpenJDK Runtime Environment (build 11.0.16+8-post-Ubuntu-0ubuntu118.04)\n",
            "OpenJDK 64-Bit Server VM (build 11.0.16+8-post-Ubuntu-0ubuntu118.04, mixed mode, sharing)\n"
          ]
        }
      ],
      "source": [
        "!java -version"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8bpgwFgcbgRd",
        "outputId": "b561d4c3-01a0-452c-a3e4-7093ffa3a4c1"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2022-11-02 23:29:57--  https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.9.zip\n",
            "Resolving www.bioinformatics.babraham.ac.uk (www.bioinformatics.babraham.ac.uk)... 149.155.133.4\n",
            "Connecting to www.bioinformatics.babraham.ac.uk (www.bioinformatics.babraham.ac.uk)|149.155.133.4|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 10249221 (9.8M) [application/zip]\n",
            "Saving to: ‘fastqc_v0.11.9.zip’\n",
            "\n",
            "fastqc_v0.11.9.zip  100%[===================>]   9.77M  1.70MB/s    in 6.2s    \n",
            "\n",
            "2022-11-02 23:30:04 (1.59 MB/s) - ‘fastqc_v0.11.9.zip’ saved [10249221/10249221]\n",
            "\n"
          ]
        }
      ],
      "source": [
        "!wget https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.9.zip"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "yTcTjOTJblcm",
        "outputId": "32bd6b8f-2e1d-40ff-e9a5-87f10fd15802"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Archive:  fastqc_v0.11.9.zip\n",
            "  inflating: FastQC/cisd-jhdf5.jar   \n",
            "   creating: FastQC/Configuration/\n",
            "  inflating: FastQC/Configuration/adapter_list.txt  \n",
            "  inflating: FastQC/Configuration/contaminant_list.txt  \n",
            "  inflating: FastQC/Configuration/limits.txt  \n",
            "  inflating: FastQC/fastqc           \n",
            "  inflating: FastQC/fastqc_icon.ico  \n",
            "   creating: FastQC/Help/\n",
            "   creating: FastQC/Help/1 Introduction/\n",
            "   creating: FastQC/Help/1 Introduction/.svn/\n",
            "  inflating: FastQC/Help/1 Introduction/.svn/entries  \n",
            "   creating: FastQC/Help/1 Introduction/.svn/props/\n",
            "   creating: FastQC/Help/1 Introduction/.svn/text-base/\n",
            "  inflating: FastQC/Help/1 Introduction/.svn/text-base/1.1 What is FastQC.html.svn-base  \n",
            "   creating: FastQC/Help/1 Introduction/.svn/tmp/\n",
            "   creating: FastQC/Help/1 Introduction/.svn/tmp/props/\n",
            "  inflating: FastQC/Help/1 Introduction/1.1 What is FastQC.html  \n",
            "   creating: FastQC/Help/2 Basic Operations/\n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/\n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/entries  \n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/props/\n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/text-base/\n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/text-base/2.1 Opening a sequence file.html.svn-base  \n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/text-base/2.2 Evaluating Results.html.svn-base  \n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/text-base/2.3 Saving a Report.html.svn-base  \n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/tmp/\n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/tmp/props/\n",
            "  inflating: FastQC/Help/2 Basic Operations/2.1 Opening a sequence file.html  \n",
            "  inflating: FastQC/Help/2 Basic Operations/2.2 Evaluating Results.html  \n",
            "  inflating: FastQC/Help/2 Basic Operations/2.3 Saving a Report.html  \n",
            "   creating: FastQC/Help/3 Analysis Modules/\n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/\n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/entries  \n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/prop-base/\n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/duplication_levels.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/prop-base/kmer_profiles.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_gc_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_n_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_quality.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_sequence_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_sequence_gc_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_sequence_quality.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_tile_quality.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/sequence_length_distribution.png.svn-base  \n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/props/\n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/text-base/\n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/1 Basic Statistics.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/10 Adapter Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/11 Kmer Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/12 Per Tile Sequence Quality.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/2 Per Base Sequence Quality.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/3 Per Sequence Quality Scores.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/4 Per Base Sequence Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/5 Per Sequence GC Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/6 Per Base N Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/7 Sequence Length Distribution.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/8 Duplicate Sequences.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/9 Overrepresented Sequences.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/duplication_levels.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/kmer_profiles.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_gc_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_n_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_quality.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_sequence_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_sequence_gc_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_sequence_quality.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_tile_quality.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/sequence_length_distribution.png.svn-base  \n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/tmp/\n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/tmp/props/\n",
            "  inflating: FastQC/Help/3 Analysis Modules/1 Basic Statistics.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/10 Adapter Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/11 Kmer Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/12 Per Tile Sequence Quality.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/2 Per Base Sequence Quality.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/3 Per Sequence Quality Scores.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/4 Per Base Sequence Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/5 Per Sequence GC Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/6 Per Base N Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/7 Sequence Length Distribution.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/8 Duplicate Sequences.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/9 Overrepresented Sequences.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/duplication_levels.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/kmer_profiles.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_gc_content.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_n_content.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_quality.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_sequence_content.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_sequence_gc_content.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_sequence_quality.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_tile_quality.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/sequence_length_distribution.png  \n",
            "  inflating: FastQC/INSTALL.txt      \n",
            "  inflating: FastQC/jbzip2-0.9.jar   \n",
            "  inflating: FastQC/LICENSE          \n",
            "  inflating: FastQC/LICENSE.txt      \n",
            "  inflating: FastQC/LICENSE_JHDF5.txt  \n",
            "   creating: FastQC/net/\n",
            "   creating: FastQC/net/sourceforge/\n",
            "   creating: FastQC/net/sourceforge/iharder/\n",
            "   creating: FastQC/net/sourceforge/iharder/base64/\n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64$1.class  \n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64$InputStream.class  \n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64$OutputStream.class  \n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64.class  \n",
            "   creating: FastQC/org/\n",
            "   creating: FastQC/org/apache/\n",
            "   creating: FastQC/org/apache/commons/\n",
            "   creating: FastQC/org/apache/commons/math3/\n",
            "   creating: FastQC/org/apache/commons/math3/analysis/\n",
            "   creating: FastQC/org/apache/commons/math3/analysis/solvers/\n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/AbstractUnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/AllowedSolution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BaseAbstractUnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BaseUnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BracketedUnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BrentSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/UnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/UnivariateSolverUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/UnivariateFunction.class  \n",
            "   creating: FastQC/org/apache/commons/math3/distribution/\n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/AbstractIntegerDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/AbstractRealDistribution$1.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/AbstractRealDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/BetaDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/BinomialDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/CauchyDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/ChiSquaredDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/FDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/GammaDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/HypergeometricDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/IntegerDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/NormalDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/PascalDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/PoissonDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/RealDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/SaddlePointExpansion.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/TDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/WeibullDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/ZipfDistribution.class  \n",
            "   creating: FastQC/org/apache/commons/math3/exception/\n",
            "  inflating: FastQC/org/apache/commons/math3/exception/ConvergenceException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/DimensionMismatchException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathArithmeticException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathIllegalArgumentException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathIllegalNumberException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathIllegalStateException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathInternalError.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MaxCountExceededException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NoBracketingException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NotFiniteNumberException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NotPositiveException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NotStrictlyPositiveException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NullArgumentException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NumberIsTooLargeException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NumberIsTooSmallException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/OutOfRangeException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/TooManyEvaluationsException.class  \n",
            "   creating: FastQC/org/apache/commons/math3/exception/util/\n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/ArgUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/ExceptionContext.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/ExceptionContextProvider.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/Localizable.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/LocalizedFormats.class  \n",
            "   creating: FastQC/org/apache/commons/math3/random/\n",
            "  inflating: FastQC/org/apache/commons/math3/random/AbstractWell.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/BitsStreamGenerator.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/RandomData.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/RandomDataImpl.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/RandomGenerator.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/Well19937c.class  \n",
            "   creating: FastQC/org/apache/commons/math3/special/\n",
            "  inflating: FastQC/org/apache/commons/math3/special/Beta$1.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Beta.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Erf.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Gamma$1.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Gamma.class  \n",
            "   creating: FastQC/org/apache/commons/math3/util/\n",
            "  inflating: FastQC/org/apache/commons/math3/util/ArithmeticUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/ContinuedFraction.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/DoubleArray.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath$ExpFracTable.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath$ExpIntTable.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath$lnMant.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMathCalc.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMathLiteralArrays.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/Incrementor$1.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/Incrementor$MaxCountExceededCallback.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/Incrementor.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/MathUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/Precision.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/ResizableDoubleArray.class  \n",
            "  inflating: FastQC/README.md        \n",
            "  inflating: FastQC/README.txt       \n",
            "  inflating: FastQC/RELEASE_NOTES.txt  \n",
            "  inflating: FastQC/run_fastqc.bat   \n",
            "  inflating: FastQC/sam-1.103.jar    \n",
            "   creating: FastQC/Templates/\n",
            "  inflating: FastQC/Templates/fastqc2fo.xsl  \n",
            "  inflating: FastQC/Templates/header_template.html  \n",
            "   creating: FastQC/Templates/Icons/\n",
            " extracting: FastQC/Templates/Icons/error.png  \n",
            " extracting: FastQC/Templates/Icons/fastqc_icon.png  \n",
            " extracting: FastQC/Templates/Icons/tick.png  \n",
            " extracting: FastQC/Templates/Icons/warning.png  \n",
            "   creating: FastQC/uk/\n",
            "   creating: FastQC/uk/ac/\n",
            "   creating: FastQC/uk/ac/babraham/\n",
            "   creating: FastQC/uk/ac/babraham/FastQC/\n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Analysis/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/AnalysisListener.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/AnalysisQueue.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/AnalysisRunner.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/OfflineRunner.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Dialogs/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/AboutDialog$1.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/AboutDialog.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/FastQCTitlePanel$SmoothJLabel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/FastQCTitlePanel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/WelcomePanel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCApplication$1.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCApplication.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCConfig.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCMenuBar.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/FileFilters/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/BAMFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/CasavaFastQFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/FastQFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/GobyFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/MappedBAMFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/SequenceFileFilter.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Graphs/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/BaseGroup.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/LineGraph.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/QualityBoxPlot.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/TileGraph.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Help/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpDialog.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpIndexRoot$FileSorter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpIndexRoot.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpPage.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpPageDisplay$HelpEditor.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpPageDisplay.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpSearchPanel.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Modules/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AbstractQCModule.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AdapterContent$Adapter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AdapterContent$ResultsTable.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AdapterContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/BasicStats$ResultsTable.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/BasicStats.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/DuplicationLevel.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Modules/GCModel/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/GCModel/GCModel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/GCModel/GCModelValue.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/KmerContent$Kmer.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/KmerContent$ResultsTable.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/KmerContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/ModuleConfig.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/ModuleFactory.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/NContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/OverRepresentedSeqs$OverrepresentedSeq.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/OverRepresentedSeqs$ResultsTable.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/OverRepresentedSeqs.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerBaseQualityScores.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerBaseSequenceContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerSequenceGCContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerSequenceQualityScores.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerTileQualityScores.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/QCModule.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/SequenceLengthDistribution.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Report/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Report/HTMLReportArchive.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Report/stylesheet.txt  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Resources/\n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/error.png  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/fastqc_icon.png  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Resources/fastqc_icon.svg  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/fastqc_icon_100.png  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/tick.png  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/warning.png  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Results/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Results/ResultsPanel$ModuleRenderer.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Results/ResultsPanel.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Sequence/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/BAMFile.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/Contaminant.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/ContaminantHit.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/ContaminentFinder.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Fast5File.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/FastQFile.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Sequence/QualityEncoding/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/QualityEncoding/PhredEncoding.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Sequence.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFactory.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFile.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFileGroup.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFormatException.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Statistics/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Statistics/NormalDistribution.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Statistics/PearsonCorrelation.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Utilities/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/CasavaBasename.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/HotColdColourGradient.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/ImageToBase64.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/MultiMemberGZIPInputStream.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/NameFormatException.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/NanoporeBasename.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/QualityCount.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/RGB.class  \n"
          ]
        }
      ],
      "source": [
        "!unzip fastqc_v0.11.9.zip"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "S79LH0aXbvLx"
      },
      "outputs": [],
      "source": [
        "!chmod a+x FastQC/fastqc"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PsLuZDt4b7jT",
        "outputId": "db70b4b7-9d12-4273-ad1d-08c925ba009a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "            FastQC - A high throughput sequence QC analysis tool\n",
            "\n",
            "SYNOPSIS\n",
            "\n",
            "\tfastqc seqfile1 seqfile2 .. seqfileN\n",
            "\n",
            "    fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam] \n",
            "           [-c contaminant file] seqfile1 .. seqfileN\n",
            "\n",
            "DESCRIPTION\n",
            "\n",
            "    FastQC reads a set of sequence files and produces from each one a quality\n",
            "    control report consisting of a number of different modules, each one of \n",
            "    which will help to identify a different potential type of problem in your\n",
            "    data.\n",
            "    \n",
            "    If no files to process are specified on the command line then the program\n",
            "    will start as an interactive graphical application.  If files are provided\n",
            "    on the command line then the program will run with no user interaction\n",
            "    required.  In this mode it is suitable for inclusion into a standardised\n",
            "    analysis pipeline.\n",
            "    \n",
            "    The options for the program as as follows:\n",
            "    \n",
            "    -h --help       Print this help file and exit\n",
            "    \n",
            "    -v --version    Print the version of the program and exit\n",
            "    \n",
            "    -o --outdir     Create all output files in the specified output directory.\n",
            "                    Please note that this directory must exist as the program\n",
            "                    will not create it.  If this option is not set then the \n",
            "                    output file for each sequence file is created in the same\n",
            "                    directory as the sequence file which was processed.\n",
            "                    \n",
            "    --casava        Files come from raw casava output. Files in the same sample\n",
            "                    group (differing only by the group number) will be analysed\n",
            "                    as a set rather than individually. Sequences with the filter\n",
            "                    flag set in the header will be excluded from the analysis.\n",
            "                    Files must have the same names given to them by casava\n",
            "                    (including being gzipped and ending with .gz) otherwise they\n",
            "                    won't be grouped together correctly.\n",
            "                    \n",
            "    --nano          Files come from nanopore sequences and are in fast5 format. In\n",
            "                    this mode you can pass in directories to process and the program\n",
            "                    will take in all fast5 files within those directories and produce\n",
            "                    a single output file from the sequences found in all files.                    \n",
            "                    \n",
            "    --nofilter      If running with --casava then don't remove read flagged by\n",
            "                    casava as poor quality when performing the QC analysis.\n",
            "                   \n",
            "    --extract       If set then the zipped output file will be uncompressed in\n",
            "                    the same directory after it has been created.  By default\n",
            "                    this option will be set if fastqc is run in non-interactive\n",
            "                    mode.\n",
            "                    \n",
            "    -j --java       Provides the full path to the java binary you want to use to\n",
            "                    launch fastqc. If not supplied then java is assumed to be in\n",
            "                    your path.\n",
            "                   \n",
            "    --noextract     Do not uncompress the output file after creating it.  You\n",
            "                    should set this option if you do not wish to uncompress\n",
            "                    the output when running in non-interactive mode.\n",
            "                    \n",
            "    --nogroup       Disable grouping of bases for reads >50bp. All reports will\n",
            "                    show data for every base in the read.  WARNING: Using this\n",
            "                    option will cause fastqc to crash and burn if you use it on\n",
            "                    really long reads, and your plots may end up a ridiculous size.\n",
            "                    You have been warned!\n",
            "                    \n",
            "    --min_length    Sets an artificial lower limit on the length of the sequence\n",
            "                    to be shown in the report.  As long as you set this to a value\n",
            "                    greater or equal to your longest read length then this will be\n",
            "                    the sequence length used to create your read groups.  This can\n",
            "                    be useful for making directly comaparable statistics from \n",
            "                    datasets with somewhat variable read lengths.\n",
            "                    \n",
            "    -f --format     Bypasses the normal sequence file format detection and\n",
            "                    forces the program to use the specified format.  Valid\n",
            "                    formats are bam,sam,bam_mapped,sam_mapped and fastq\n",
            "                    \n",
            "    -t --threads    Specifies the number of files which can be processed\n",
            "                    simultaneously.  Each thread will be allocated 250MB of\n",
            "                    memory so you shouldn't run more threads than your\n",
            "                    available memory will cope with, and not more than\n",
            "                    6 threads on a 32 bit machine\n",
            "                  \n",
            "    -c              Specifies a non-default file which contains the list of\n",
            "    --contaminants  contaminants to screen overrepresented sequences against.\n",
            "                    The file must contain sets of named contaminants in the\n",
            "                    form name[tab]sequence.  Lines prefixed with a hash will\n",
            "                    be ignored.\n",
            "\n",
            "    -a              Specifies a non-default file which contains the list of\n",
            "    --adapters      adapter sequences which will be explicity searched against\n",
            "                    the library. The file must contain sets of named adapters\n",
            "                    in the form name[tab]sequence.  Lines prefixed with a hash\n",
            "                    will be ignored.\n",
            "                    \n",
            "    -l              Specifies a non-default file which contains a set of criteria\n",
            "    --limits        which will be used to determine the warn/error limits for the\n",
            "                    various modules.  This file can also be used to selectively \n",
            "                    remove some modules from the output all together.  The format\n",
            "                    needs to mirror the default limits.txt file found in the\n",
            "                    Configuration folder.\n",
            "                    \n",
            "   -k --kmers       Specifies the length of Kmer to look for in the Kmer content\n",
            "                    module. Specified Kmer length must be between 2 and 10. Default\n",
            "                    length is 7 if not specified.\n",
            "                    \n",
            "   -q --quiet       Supress all progress messages on stdout and only report errors.\n",
            "   \n",
            "   -d --dir         Selects a directory to be used for temporary files written when\n",
            "                    generating report images. Defaults to system temp directory if\n",
            "                    not specified.\n",
            "                    \n",
            "BUGS\n",
            "\n",
            "    Any bugs in fastqc should be reported either to simon.andrews@babraham.ac.uk\n",
            "    or in www.bioinformatics.babraham.ac.uk/bugzilla/\n",
            "                   \n",
            "    \n"
          ]
        }
      ],
      "source": [
        "!./FastQC/fastqc --help"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XUZS4GUkHIEG"
      },
      "source": [
        "## Установка других необходимых пакетов\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "lrtRbP3fAvj8"
      },
      "outputs": [],
      "source": [
        "pip install -q condacolab"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "jQ1Oqzb9BXoG"
      },
      "outputs": [],
      "source": [
        "import condacolab"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "q5FUOHCgBZWp",
        "outputId": "1a538fcb-341d-4ef1-ed29-be1465680b22"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "⏬ Downloading https://github.com/jaimergp/miniforge/releases/latest/download/Mambaforge-colab-Linux-x86_64.sh...\n",
            "📦 Installing...\n",
            "📌 Adjusting configuration...\n",
            "🩹 Patching environment...\n",
            "⏲ Done in 0:00:21\n",
            "🔁 Restarting kernel...\n"
          ]
        }
      ],
      "source": [
        "condacolab.install()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "sCBV_31qHOik",
        "outputId": "fd884b30-2d43-4d3f-81a4-86700aa89461"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting package metadata (current_repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "Solving environment: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  added / updated specs:\n",
            "    - trimmomatic\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    ca-certificates-2022.9.24  |       ha878542_0         150 KB  conda-forge\n",
            "    certifi-2022.9.24          |     pyhd8ed1ab_0         155 KB  conda-forge\n",
            "    conda-22.9.0               |   py37h89c1867_1         960 KB  conda-forge\n",
            "    openjdk-8.0.332            |       h166bdaf_0        97.8 MB  conda-forge\n",
            "    openssl-1.1.1s             |       h166bdaf_0         2.1 MB  conda-forge\n",
            "    trimmomatic-0.39           |       hdfd78af_2         144 KB  bioconda\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:       101.3 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  openjdk            conda-forge/linux-64::openjdk-8.0.332-h166bdaf_0\n",
            "  trimmomatic        bioconda/noarch::trimmomatic-0.39-hdfd78af_2\n",
            "\n",
            "The following packages will be UPDATED:\n",
            "\n",
            "  ca-certificates                      2022.6.15-ha878542_0 --> 2022.9.24-ha878542_0\n",
            "  certifi            conda-forge/linux-64::certifi-2022.6.~ --> conda-forge/noarch::certifi-2022.9.24-pyhd8ed1ab_0\n",
            "  conda                               4.14.0-py37h89c1867_0 --> 22.9.0-py37h89c1867_1\n",
            "  openssl                                 1.1.1q-h166bdaf_0 --> 1.1.1s-h166bdaf_0\n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages\n",
            "certifi-2022.9.24    | 155 KB    | : 100% 1.0/1 [00:00<00:00, 10.51it/s]\n",
            "openjdk-8.0.332      | 97.8 MB   | : 100% 1.0/1 [00:10<00:00, 10.44s/it]               \n",
            "ca-certificates-2022 | 150 KB    | : 100% 1.0/1 [00:00<00:00, 18.78it/s]\n",
            "conda-22.9.0         | 960 KB    | : 100% 1.0/1 [00:00<00:00,  3.71it/s]\n",
            "openssl-1.1.1s       | 2.1 MB    | : 100% 1.0/1 [00:00<00:00,  3.57it/s]\n",
            "trimmomatic-0.39     | 144 KB    | : 100% 1.0/1 [00:00<00:00, 19.25it/s]\n",
            "Preparing transaction: \\ \b\bdone\n",
            "Verifying transaction: / \b\b- \b\bdone\n",
            "Executing transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Retrieving notices: ...working... done\n"
          ]
        }
      ],
      "source": [
        "!conda install -c bioconda trimmomatic"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "AUML3dvsqT9E",
        "outputId": "2a02756d-d107-483c-b855-457d77bd2755"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "The following package was automatically installed and is no longer required:\n",
            "  libnvidia-common-460\n",
            "Use 'apt autoremove' to remove it.\n",
            "The following additional packages will be installed:\n",
            "  cwltool libc-ares2 libhts2 libjs-bootstrap nodejs nodejs-doc\n",
            "  python-asn1crypto python-avro python-cachecontrol python-certifi\n",
            "  python-cffi-backend python-chardet python-cryptography python-enum34\n",
            "  python-html5lib python-idna python-ipaddress python-isodate python-lockfile\n",
            "  python-mistune python-openssl python-pkg-resources python-pyparsing\n",
            "  python-rdflib python-rdflib-jsonld python-requests python-ruamel.yaml\n",
            "  python-schema-salad python-shellescape python-six python-sparqlwrapper\n",
            "  python-typing python-urllib3 python-webencodings\n",
            "Suggested packages:\n",
            "  bowtie2-examples python-cryptography-doc python-cryptography-vectors\n",
            "  python-enum34-doc python-genshi python-lxml python-lockfile-doc\n",
            "  python-openssl-doc python-openssl-dbg python-setuptools python-pyparsing-doc\n",
            "  python-rdflib-doc python-rdflib-tools python-socks python-ntlm\n",
            "The following NEW packages will be installed:\n",
            "  bowtie2 cwltool libc-ares2 libhts2 libjs-bootstrap nodejs nodejs-doc\n",
            "  python-asn1crypto python-avro python-cachecontrol python-certifi\n",
            "  python-cffi-backend python-chardet python-cryptography python-enum34\n",
            "  python-html5lib python-idna python-ipaddress python-isodate python-lockfile\n",
            "  python-mistune python-openssl python-pkg-resources python-pyparsing\n",
            "  python-rdflib python-rdflib-jsonld python-requests python-ruamel.yaml\n",
            "  python-schema-salad python-shellescape python-six python-sparqlwrapper\n",
            "  python-typing python-urllib3 python-webencodings samtools\n",
            "0 upgraded, 36 newly installed, 0 to remove and 4 not upgraded.\n",
            "Need to get 9,743 kB of archives.\n",
            "After this operation, 46.1 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-pkg-resources all 39.0.1-2 [128 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-six all 1.11.0-2 [11.3 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-isodate all 0.6.0-1 [23.8 kB]\n",
            "Get:4 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-pyparsing all 2.2.0+dfsg1-2 [52.1 kB]\n",
            "Get:5 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-rdflib all 4.2.1-2 [237 kB]\n",
            "Get:6 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-certifi all 2018.1.18-2 [144 kB]\n",
            "Get:7 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-chardet all 3.0.4-1 [80.3 kB]\n",
            "Get:8 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-idna all 2.6-1 [32.4 kB]\n",
            "Get:9 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 python-urllib3 all 1.22-1ubuntu0.18.04.2 [86.0 kB]\n",
            "Get:10 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 python-requests all 2.18.4-2ubuntu0.1 [58.5 kB]\n",
            "Get:11 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-ruamel.yaml amd64 0.15.34-1 [183 kB]\n",
            "Get:12 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-cachecontrol all 0.11.7-1 [15.2 kB]\n",
            "Get:13 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-lockfile all 1:0.12.2-2 [14.6 kB]\n",
            "Get:14 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-mistune all 0.8.3-2 [13.8 kB]\n",
            "Get:15 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-rdflib-jsonld all 0.4.0-3 [12.2 kB]\n",
            "Get:16 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-typing all 3.6.2-1 [22.2 kB]\n",
            "Get:17 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-avro all 1.8.2+dfsg-1 [23.2 kB]\n",
            "Get:18 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libjs-bootstrap all 3.3.7+dfsg-2 [228 kB]\n",
            "Get:19 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-schema-salad all 2.6.20171201034858-3 [260 kB]\n",
            "Get:20 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-shellescape all 3.4.1-2 [4,432 B]\n",
            "Get:21 http://archive.ubuntu.com/ubuntu bionic/universe amd64 cwltool all 1.0.20180302231433-1 [143 kB]\n",
            "Get:22 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libhts2 amd64 1.7-2 [300 kB]\n",
            "Get:23 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 nodejs-doc all 8.10.0~dfsg-2ubuntu0.4 [752 kB]\n",
            "Get:24 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-asn1crypto all 0.24.0-1 [72.7 kB]\n",
            "Get:25 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-cffi-backend amd64 1.11.5-1 [63.4 kB]\n",
            "Get:26 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-enum34 all 1.1.6-2 [34.8 kB]\n",
            "Get:27 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-ipaddress all 1.0.17-1 [18.2 kB]\n",
            "Get:28 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 python-cryptography amd64 2.1.4-1ubuntu1.4 [276 kB]\n",
            "Get:29 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-webencodings all 0.5-2 [10.3 kB]\n",
            "Get:30 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-html5lib all 0.999999999-1 [83.6 kB]\n",
            "Get:31 http://archive.ubuntu.com/ubuntu bionic/main amd64 python-openssl all 17.5.0-1ubuntu1 [41.3 kB]\n",
            "Get:32 http://archive.ubuntu.com/ubuntu bionic/universe amd64 python-sparqlwrapper all 1.7.6-2 [21.8 kB]\n",
            "Get:33 http://archive.ubuntu.com/ubuntu bionic/universe amd64 samtools amd64 1.7-1 [288 kB]\n",
            "Get:34 http://archive.ubuntu.com/ubuntu bionic/universe amd64 bowtie2 amd64 2.3.4.1-1 [1,174 kB]\n",
            "Get:35 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 libc-ares2 amd64 1.14.0-1ubuntu0.1 [37.5 kB]\n",
            "Get:36 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 nodejs amd64 8.10.0~dfsg-2ubuntu0.4 [4,796 kB]\n",
            "Fetched 9,743 kB in 2s (4,216 kB/s)\n",
            "Extracting templates from packages: 100%\n",
            "Selecting previously unselected package python-pkg-resources.\n",
            "(Reading database ... 123942 files and directories currently installed.)\n",
            "Preparing to unpack .../00-python-pkg-resources_39.0.1-2_all.deb ...\n",
            "Unpacking python-pkg-resources (39.0.1-2) ...\n",
            "Selecting previously unselected package python-six.\n",
            "Preparing to unpack .../01-python-six_1.11.0-2_all.deb ...\n",
            "Unpacking python-six (1.11.0-2) ...\n",
            "Selecting previously unselected package python-isodate.\n",
            "Preparing to unpack .../02-python-isodate_0.6.0-1_all.deb ...\n",
            "Unpacking python-isodate (0.6.0-1) ...\n",
            "Selecting previously unselected package python-pyparsing.\n",
            "Preparing to unpack .../03-python-pyparsing_2.2.0+dfsg1-2_all.deb ...\n",
            "Unpacking python-pyparsing (2.2.0+dfsg1-2) ...\n",
            "Selecting previously unselected package python-rdflib.\n",
            "Preparing to unpack .../04-python-rdflib_4.2.1-2_all.deb ...\n",
            "Unpacking python-rdflib (4.2.1-2) ...\n",
            "Selecting previously unselected package python-certifi.\n",
            "Preparing to unpack .../05-python-certifi_2018.1.18-2_all.deb ...\n",
            "Unpacking python-certifi (2018.1.18-2) ...\n",
            "Selecting previously unselected package python-chardet.\n",
            "Preparing to unpack .../06-python-chardet_3.0.4-1_all.deb ...\n",
            "Unpacking python-chardet (3.0.4-1) ...\n",
            "Selecting previously unselected package python-idna.\n",
            "Preparing to unpack .../07-python-idna_2.6-1_all.deb ...\n",
            "Unpacking python-idna (2.6-1) ...\n",
            "Selecting previously unselected package python-urllib3.\n",
            "Preparing to unpack .../08-python-urllib3_1.22-1ubuntu0.18.04.2_all.deb ...\n",
            "Unpacking python-urllib3 (1.22-1ubuntu0.18.04.2) ...\n",
            "Selecting previously unselected package python-requests.\n",
            "Preparing to unpack .../09-python-requests_2.18.4-2ubuntu0.1_all.deb ...\n",
            "Unpacking python-requests (2.18.4-2ubuntu0.1) ...\n",
            "Selecting previously unselected package python-ruamel.yaml.\n",
            "Preparing to unpack .../10-python-ruamel.yaml_0.15.34-1_amd64.deb ...\n",
            "Unpacking python-ruamel.yaml (0.15.34-1) ...\n",
            "Selecting previously unselected package python-cachecontrol.\n",
            "Preparing to unpack .../11-python-cachecontrol_0.11.7-1_all.deb ...\n",
            "Unpacking python-cachecontrol (0.11.7-1) ...\n",
            "Selecting previously unselected package python-lockfile.\n",
            "Preparing to unpack .../12-python-lockfile_1%3a0.12.2-2_all.deb ...\n",
            "Unpacking python-lockfile (1:0.12.2-2) ...\n",
            "Selecting previously unselected package python-mistune.\n",
            "Preparing to unpack .../13-python-mistune_0.8.3-2_all.deb ...\n",
            "Unpacking python-mistune (0.8.3-2) ...\n",
            "Selecting previously unselected package python-rdflib-jsonld.\n",
            "Preparing to unpack .../14-python-rdflib-jsonld_0.4.0-3_all.deb ...\n",
            "Unpacking python-rdflib-jsonld (0.4.0-3) ...\n",
            "Selecting previously unselected package python-typing.\n",
            "Preparing to unpack .../15-python-typing_3.6.2-1_all.deb ...\n",
            "Unpacking python-typing (3.6.2-1) ...\n",
            "Selecting previously unselected package python-avro.\n",
            "Preparing to unpack .../16-python-avro_1.8.2+dfsg-1_all.deb ...\n",
            "Unpacking python-avro (1.8.2+dfsg-1) ...\n",
            "Selecting previously unselected package libjs-bootstrap.\n",
            "Preparing to unpack .../17-libjs-bootstrap_3.3.7+dfsg-2_all.deb ...\n",
            "Unpacking libjs-bootstrap (3.3.7+dfsg-2) ...\n",
            "Selecting previously unselected package python-schema-salad.\n",
            "Preparing to unpack .../18-python-schema-salad_2.6.20171201034858-3_all.deb ...\n",
            "Unpacking python-schema-salad (2.6.20171201034858-3) ...\n",
            "Selecting previously unselected package python-shellescape.\n",
            "Preparing to unpack .../19-python-shellescape_3.4.1-2_all.deb ...\n",
            "Unpacking python-shellescape (3.4.1-2) ...\n",
            "Selecting previously unselected package cwltool.\n",
            "Preparing to unpack .../20-cwltool_1.0.20180302231433-1_all.deb ...\n",
            "Unpacking cwltool (1.0.20180302231433-1) ...\n",
            "Selecting previously unselected package libhts2:amd64.\n",
            "Preparing to unpack .../21-libhts2_1.7-2_amd64.deb ...\n",
            "Unpacking libhts2:amd64 (1.7-2) ...\n",
            "Selecting previously unselected package nodejs-doc.\n",
            "Preparing to unpack .../22-nodejs-doc_8.10.0~dfsg-2ubuntu0.4_all.deb ...\n",
            "Unpacking nodejs-doc (8.10.0~dfsg-2ubuntu0.4) ...\n",
            "Selecting previously unselected package python-asn1crypto.\n",
            "Preparing to unpack .../23-python-asn1crypto_0.24.0-1_all.deb ...\n",
            "Unpacking python-asn1crypto (0.24.0-1) ...\n",
            "Selecting previously unselected package python-cffi-backend.\n",
            "Preparing to unpack .../24-python-cffi-backend_1.11.5-1_amd64.deb ...\n",
            "Unpacking python-cffi-backend (1.11.5-1) ...\n",
            "Selecting previously unselected package python-enum34.\n",
            "Preparing to unpack .../25-python-enum34_1.1.6-2_all.deb ...\n",
            "Unpacking python-enum34 (1.1.6-2) ...\n",
            "Selecting previously unselected package python-ipaddress.\n",
            "Preparing to unpack .../26-python-ipaddress_1.0.17-1_all.deb ...\n",
            "Unpacking python-ipaddress (1.0.17-1) ...\n",
            "Selecting previously unselected package python-cryptography.\n",
            "Preparing to unpack .../27-python-cryptography_2.1.4-1ubuntu1.4_amd64.deb ...\n",
            "Unpacking python-cryptography (2.1.4-1ubuntu1.4) ...\n",
            "Selecting previously unselected package python-webencodings.\n",
            "Preparing to unpack .../28-python-webencodings_0.5-2_all.deb ...\n",
            "Unpacking python-webencodings (0.5-2) ...\n",
            "Selecting previously unselected package python-html5lib.\n",
            "Preparing to unpack .../29-python-html5lib_0.999999999-1_all.deb ...\n",
            "Unpacking python-html5lib (0.999999999-1) ...\n",
            "Selecting previously unselected package python-openssl.\n",
            "Preparing to unpack .../30-python-openssl_17.5.0-1ubuntu1_all.deb ...\n",
            "Unpacking python-openssl (17.5.0-1ubuntu1) ...\n",
            "Selecting previously unselected package python-sparqlwrapper.\n",
            "Preparing to unpack .../31-python-sparqlwrapper_1.7.6-2_all.deb ...\n",
            "Unpacking python-sparqlwrapper (1.7.6-2) ...\n",
            "Selecting previously unselected package samtools.\n",
            "Preparing to unpack .../32-samtools_1.7-1_amd64.deb ...\n",
            "Unpacking samtools (1.7-1) ...\n",
            "Selecting previously unselected package bowtie2.\n",
            "Preparing to unpack .../33-bowtie2_2.3.4.1-1_amd64.deb ...\n",
            "Unpacking bowtie2 (2.3.4.1-1) ...\n",
            "Selecting previously unselected package libc-ares2:amd64.\n",
            "Preparing to unpack .../34-libc-ares2_1.14.0-1ubuntu0.1_amd64.deb ...\n",
            "Unpacking libc-ares2:amd64 (1.14.0-1ubuntu0.1) ...\n",
            "Selecting previously unselected package nodejs.\n",
            "Preparing to unpack .../35-nodejs_8.10.0~dfsg-2ubuntu0.4_amd64.deb ...\n",
            "Unpacking nodejs (8.10.0~dfsg-2ubuntu0.4) ...\n",
            "Setting up python-idna (2.6-1) ...\n",
            "Setting up nodejs-doc (8.10.0~dfsg-2ubuntu0.4) ...\n",
            "Setting up python-asn1crypto (0.24.0-1) ...\n",
            "Setting up python-mistune (0.8.3-2) ...\n",
            "Setting up python-certifi (2018.1.18-2) ...\n",
            "Setting up python-avro (1.8.2+dfsg-1) ...\n",
            "Setting up python-pkg-resources (39.0.1-2) ...\n",
            "Setting up bowtie2 (2.3.4.1-1) ...\n",
            "Setting up python-ruamel.yaml (0.15.34-1) ...\n",
            "Setting up python-cffi-backend (1.11.5-1) ...\n",
            "Setting up python-six (1.11.0-2) ...\n",
            "Setting up python-isodate (0.6.0-1) ...\n",
            "Setting up python-shellescape (3.4.1-2) ...\n",
            "Setting up python-enum34 (1.1.6-2) ...\n",
            "Setting up libjs-bootstrap (3.3.7+dfsg-2) ...\n",
            "Setting up python-lockfile (1:0.12.2-2) ...\n",
            "Setting up python-ipaddress (1.0.17-1) ...\n",
            "Setting up libc-ares2:amd64 (1.14.0-1ubuntu0.1) ...\n",
            "Setting up libhts2:amd64 (1.7-2) ...\n",
            "Setting up python-webencodings (0.5-2) ...\n",
            "Setting up python-typing (3.6.2-1) ...\n",
            "Setting up python-pyparsing (2.2.0+dfsg1-2) ...\n",
            "Setting up python-urllib3 (1.22-1ubuntu0.18.04.2) ...\n",
            "Setting up python-chardet (3.0.4-1) ...\n",
            "Setting up nodejs (8.10.0~dfsg-2ubuntu0.4) ...\n",
            "update-alternatives: using /usr/bin/nodejs to provide /usr/bin/js (js) in auto mode\n",
            "Setting up python-rdflib (4.2.1-2) ...\n",
            "Setting up python-sparqlwrapper (1.7.6-2) ...\n",
            "Setting up python-rdflib-jsonld (0.4.0-3) ...\n",
            "Setting up python-cryptography (2.1.4-1ubuntu1.4) ...\n",
            "Setting up python-requests (2.18.4-2ubuntu0.1) ...\n",
            "Setting up python-cachecontrol (0.11.7-1) ...\n",
            "Setting up python-schema-salad (2.6.20171201034858-3) ...\n",
            "Setting up python-html5lib (0.999999999-1) ...\n",
            "Setting up samtools (1.7-1) ...\n",
            "Setting up python-openssl (17.5.0-1ubuntu1) ...\n",
            "Setting up cwltool (1.0.20180302231433-1) ...\n",
            "Processing triggers for libc-bin (2.27-3ubuntu1.6) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/ is not a symbolic link\n",
            "\n",
            "Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\n"
          ]
        }
      ],
      "source": [
        "!apt-get install bowtie2 samtools"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-z4GQiMfU1wf"
      },
      "source": [
        "## Загрузка данных и запуск FastQC"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "lFWnMg0uU0sJ",
        "outputId": "2cfdd693-f1c7-4bad-c156-6666a0e76a54"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2022-11-02 23:31:42--  https://figshare.com/ndownloader/files/23769689\n",
            "Resolving figshare.com (figshare.com)... 34.249.11.40, 34.248.76.93, 2a05:d018:1f4:d000:e0da:3116:1b67:a60b, ...\n",
            "Connecting to figshare.com (figshare.com)|34.249.11.40|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/23769689/amp_res_1.fastq.gz?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20221102/eu-west-1/s3/aws4_request&X-Amz-Date=20221102T233142Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=ccecfb2f838504c60aa236810b00e449cdca11bec47c72a259aa5b50c9d67f0d [following]\n",
            "--2022-11-02 23:31:42--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/23769689/amp_res_1.fastq.gz?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20221102/eu-west-1/s3/aws4_request&X-Amz-Date=20221102T233142Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=ccecfb2f838504c60aa236810b00e449cdca11bec47c72a259aa5b50c9d67f0d\n",
            "Resolving s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)... 52.218.26.131, 52.218.122.16, 52.218.24.42, ...\n",
            "Connecting to s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)|52.218.26.131|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 44390940 (42M) [application/gzip]\n",
            "Saving to: ‘amp_res_1.fastq.gz’\n",
            "\n",
            "amp_res_1.fastq.gz  100%[===================>]  42.33M  14.7MB/s    in 2.9s    \n",
            "\n",
            "2022-11-02 23:31:46 (14.7 MB/s) - ‘amp_res_1.fastq.gz’ saved [44390940/44390940]\n",
            "\n",
            "--2022-11-02 23:31:46--  https://figshare.com/ndownloader/files/23769692\n",
            "Resolving figshare.com (figshare.com)... 34.249.11.40, 34.248.76.93, 2a05:d018:1f4:d000:e0da:3116:1b67:a60b, ...\n",
            "Connecting to figshare.com (figshare.com)|34.249.11.40|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/23769692/amp_res_2.fastq.gz?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20221102/eu-west-1/s3/aws4_request&X-Amz-Date=20221102T233146Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=0a123faa264acabdf6d90d8bd3fe269eb23108d4161a1b7fb2ab41d4833a5064 [following]\n",
            "--2022-11-02 23:31:46--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/23769692/amp_res_2.fastq.gz?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20221102/eu-west-1/s3/aws4_request&X-Amz-Date=20221102T233146Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=0a123faa264acabdf6d90d8bd3fe269eb23108d4161a1b7fb2ab41d4833a5064\n",
            "Resolving s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)... 52.218.26.131, 52.218.122.16, 52.218.24.42, ...\n",
            "Connecting to s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)|52.218.26.131|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 44375379 (42M) [application/gzip]\n",
            "Saving to: ‘amp_res_2.fastq.gz’\n",
            "\n",
            "amp_res_2.fastq.gz  100%[===================>]  42.32M  14.1MB/s    in 3.0s    \n",
            "\n",
            "2022-11-02 23:31:50 (14.1 MB/s) - ‘amp_res_2.fastq.gz’ saved [44375379/44375379]\n",
            "\n",
            "--2022-11-02 23:31:50--  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_cds_from_genomic.fna.gz\n",
            "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 165.112.9.229, 165.112.9.230, 2607:f220:41e:250::13, ...\n",
            "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|165.112.9.229|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1479110 (1.4M) [application/x-gzip]\n",
            "Saving to: ‘GCF_000005845.2_ASM584v2_cds_from_genomic.fna.gz’\n",
            "\n",
            "GCF_000005845.2_ASM 100%[===================>]   1.41M  4.12MB/s    in 0.3s    \n",
            "\n",
            "2022-11-02 23:31:51 (4.12 MB/s) - ‘GCF_000005845.2_ASM584v2_cds_from_genomic.fna.gz’ saved [1479110/1479110]\n",
            "\n",
            "--2022-11-02 23:31:51--  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.gff.gz\n",
            "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 165.112.9.229, 165.112.9.230, 2607:f220:41e:250::13, ...\n",
            "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|165.112.9.229|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 433547 (423K) [application/x-gzip]\n",
            "Saving to: ‘GCF_000005845.2_ASM584v2_genomic.gff.gz’\n",
            "\n",
            "GCF_000005845.2_ASM 100%[===================>] 423.39K  1.45MB/s    in 0.3s    \n",
            "\n",
            "2022-11-02 23:31:52 (1.45 MB/s) - ‘GCF_000005845.2_ASM584v2_genomic.gff.gz’ saved [433547/433547]\n",
            "\n"
          ]
        }
      ],
      "source": [
        "# resistant\n",
        "!wget 'https://figshare.com/ndownloader/files/23769689' -O amp_res_1.fastq.gz\n",
        "!wget 'https://figshare.com/ndownloader/files/23769692' -O amp_res_2.fastq.gz\n",
        "# non-resistant\n",
        "!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.fna.gz\n",
        "!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.gff.gz"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "v73fH7WiVDu9"
      },
      "outputs": [],
      "source": [
        "# распаковка ахрива\n",
        "!gzip -d /content/GCF_000005845.2_ASM584v2_genomic.fna.gz /content/GCF_000005845.2_ASM584v2_genomic.gff.gz /content/amp_res_1.fastq.gz /content/amp_res_2.fastq.gz"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!head -20 amp_res_1.fastq"
      ],
      "metadata": {
        "id": "mwM8qd3zY5CY",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "8e99f2c3-b9d0-48a7-fa11-e9f55b4553f3"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "@SRR1363257.37 GWZHISEQ01:153:C1W31ACXX:5:1101:14027:2198 length=101\n",
            "GGTTGCAGATTCGCAGTGTCGCTGTTCCAGCGCATCACATCTTTGATGTTCACGCCGTGGCGTTTAGCAATGCTTGAAAGCGAATCGCCTTTGCCCACACG\n",
            "+\n",
            "@?:=:;DBFADH;CAECEE@@E:FFHGAE4?C?DE<BFGEC>?>FHE4BFFIIFHIBABEECA83;>>@>@CCCDC9@@CC08<@?@BB@9:CC#######\n",
            "@SRR1363257.46 GWZHISEQ01:153:C1W31ACXX:5:1101:19721:2155 length=101\n",
            "GTATGAGGTTTTGCTGCATTCTCTGNGCGAATATTAACTCCNTNNNNNTTATAGTTCAAAGCAAGTACCTGTCTCTTATACACATCTCCGAGCCCACGAGC\n",
            "+\n",
            "@@<?=D?D==?<AFGDF+AIHEACH#22<:?E8??:9??GG#0#####000;CF=C)4.==CA@@@)=7?C7?E37;3@>;;(.;>AB#############\n",
            "@SRR1363257.77 GWZHISEQ01:153:C1W31ACXX:5:1101:5069:2307 length=101\n",
            "GCTTCTCTTAACTGAGGTCACCATCATGCCGTTAAGTCCCTACCTCTCTTTTGCCGGTAACTGTTCCGCCGCGATTGCCTTTTATCTGTCTCTTATACACC\n",
            "+\n",
            "??<DBD;4C2=<BB>:AC;<CF<CE@FE9@E1C@891CD*9:?:3D@DD4?D<DD:0;@A=AEIDDA##################################\n",
            "@SRR1363257.78 GWZHISEQ01:153:C1W31ACXX:5:1101:5178:2440 length=101\n",
            "GCATAAGGACGATCGCTCCAGAGTAAAATAAATACGCGCATGTGATACTCACAATACCAATGGTGAAGTTACGGGACTTAAACAAACTGAGATCAAGAATC\n",
            "+\n",
            "CCCFFFFFHHHHHJJJJJJJJJJFFHIJJJJJJJJJJJJJJJJJJJJJJJIJHHHHHHFDEDF;AEEEEEEDDDDDBBACDDDCDDDDCCDDDDDDCCDC3\n",
            "@SRR1363257.96 GWZHISEQ01:153:C1W31ACXX:5:1101:6707:2460 length=101\n",
            "TCATTAAGCCGTGGTGGATGTGCCATAGCGCACCGCAAAGTTAAGAAACCGAATATTGGGTTTAGTCTTGTTTCATAATTGTTGCAATGAAACGCGGTGAA\n",
            "+\n",
            "CCCFFFFFHHHHHJHIIJIIIIJJJJJJGIJJJJJIJJIIGHJJJJJIIJJDHFFFFFEDACDDDCDDDDCCDDECACCDCCCDACDDDDCCDDDDDBD@A\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!head -20 amp_res_2.fastq"
      ],
      "metadata": {
        "id": "PRil7tyHZIi6",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "872c3021-4b1c-49f1-d398-7e43c33b9f18"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "@SRR1363257.37 GWZHISEQ01:153:C1W31ACXX:5:1101:14027:2198 length=101\n",
            "GATCTAAGCTGAAGCCAGGCCAAAGTTTGACGATTGGTGCAGGCAGTAGCGCACAGCGACTGGCAAACAACAGCGATAGCATTACGTATCGTGTGCGCAAA\n",
            "+\n",
            "???BDB:DFHBFD@9;;+A;AFGH;ABHFHHGE@9:B:??@D>@;F?D8<<F8AA9EHHD8'..;5?A?A992(',(59CC3@C>22::A238+2>B<>B<\n",
            "@SRR1363257.46 GWZHISEQ01:153:C1W31ACXX:5:1101:19721:2155 length=101\n",
            "GTACTTGCTTTGNACTATAATATGCACGGAGNTAATATTCGCTCAGAGAATGCAGCAAAACCTCATACCTGTCTCTTATACACATCTGACGCTGCCGACGA\n",
            "+\n",
            ";@@DB?B;CFBB#2<:CB:FH<C@:<A?C::#1:86:BG9:8?8688?888EBF;783)=6-7=CC;ECD);?7;;>>AE;>(5;->AC@;B@;8?#####\n",
            "@SRR1363257.77 GWZHISEQ01:153:C1W31ACXX:5:1101:5069:2307 length=101\n",
            "ATAATAGGCAATCGCGTCGGAACAGTTACCGGCCAAAGAGAGGCAGGGACTTAACGGCATGATGGTGACCTCAGTTAAGAGAAGCCTGTCTCTTATACACA\n",
            "+\n",
            "+=?;:2,+A++AC:C:2@F6:CD:B09B?4)8@''8=))8=;=((5=4@?;@6;@?@BB;(535::>:>3(::(44:@::@3((9<32+::@(4@4+:>C3\n",
            "@SRR1363257.78 GWZHISEQ01:153:C1W31ACXX:5:1101:5178:2440 length=101\n",
            "ATATTAACAGTAGTATCAGTTATTTCTCTGATCTCTTTAGTCATTTGGGAGTCGACCTCAGAGAACCCGATTCTTGATCTCAGTTTGTTTAAGTCCCGTAA\n",
            "+\n",
            "BCCFFFFFHHHHHHIJJIJJJJJIJJJJIJGIJJJJJJIJHIHJJIJIIGGGHIJIJJJIJIJJJJJJJGHHHHHFFFFFFEEEFEEED?AACCDCCDDDB\n",
            "@SRR1363257.96 GWZHISEQ01:153:C1W31ACXX:5:1101:6707:2460 length=101\n",
            "GTTTCACCGCGTTTCATTGCAACAATTATGAAACAAGACTAAACCCAATATTCGGTTTCTTAACTTTGCGGTGCGCTATGGCACATCCACCACGGCTTAAT\n",
            "+\n",
            "CCCFFFFFHGHHHJIJJJJJJIJJJJJJIJIJJIJJIJIJJJJJJJJIJJFHIIJFIGJJJGIHHHHHGFFDDDDDDDDDDDDDDDDDDDABDDDDDDDCD\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!head -20 GCF_000005845.2_ASM584v2_genomic.fna"
      ],
      "metadata": {
        "id": "ldN74vUDZMxK",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "5a59937a-0af6-4f3e-d79d-223cd102f442"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            ">lcl|NC_000913.3_cds_NP_414542.1_1 [gene=thrL] [locus_tag=b0001] [db_xref=UniProtKB/Swiss-Prot:P0AD86] [protein=thr operon leader peptide] [protein_id=NP_414542.1] [location=190..255] [gbkey=CDS]\n",
            "ATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCACAGGTAACGGTGCGGGCTGA\n",
            ">lcl|NC_000913.3_cds_NP_414543.1_2 [gene=thrA] [locus_tag=b0002] [db_xref=UniProtKB/Swiss-Prot:P00561] [protein=fused aspartate kinase/homoserine dehydrogenase 1] [protein_id=NP_414543.1] [location=337..2799] [gbkey=CDS]\n",
            "ATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAACGTTTTCTGCGTGTTGCCGATATTCTGGAAAGCAA\n",
            "TGCCAGGCAGGGGCAGGTGGCCACCGTCCTCTCTGCCCCCGCCAAAATCACCAACCACCTGGTGGCGATGATTGAAAAAA\n",
            "CCATTAGCGGCCAGGATGCTTTACCCAATATCAGCGATGCCGAACGTATTTTTGCCGAACTTTTGACGGGACTCGCCGCC\n",
            "GCCCAGCCGGGGTTCCCGCTGGCGCAATTGAAAACTTTCGTCGATCAGGAATTTGCCCAAATAAAACATGTCCTGCATGG\n",
            "CATTAGTTTGTTGGGGCAGTGCCCGGATAGCATCAACGCTGCGCTGATTTGCCGTGGCGAGAAAATGTCGATCGCCATTA\n",
            "TGGCCGGCGTATTAGAAGCGCGCGGTCACAACGTTACTGTTATCGATCCGGTCGAAAAACTGCTGGCAGTGGGGCATTAC\n",
            "CTCGAATCTACCGTCGATATTGCTGAGTCCACCCGCCGTATTGCGGCAAGCCGCATTCCGGCTGATCACATGGTGCTGAT\n",
            "GGCAGGTTTCACCGCCGGTAATGAAAAAGGCGAACTGGTGGTGCTTGGACGCAACGGTTCCGACTACTCTGCTGCGGTGC\n",
            "TGGCTGCCTGTTTACGCGCCGATTGTTGCGAGATTTGGACGGACGTTGACGGGGTCTATACCTGCGACCCGCGTCAGGTG\n",
            "CCCGATGCGAGGTTGTTGAAGTCGATGTCCTACCAGGAAGCGATGGAGCTTTCCTACTTCGGCGCTAAAGTTCTTCACCC\n",
            "CCGCACCATTACCCCCATCGCCCAGTTCCAGATCCCTTGCCTGATTAAAAATACCGGAAATCCTCAAGCACCAGGTACGC\n",
            "TCATTGGTGCCAGCCGTGATGAAGACGAATTACCGGTCAAGGGCATTTCCAATCTGAATAACATGGCAATGTTCAGCGTT\n",
            "TCTGGTCCGGGGATGAAAGGGATGGTCGGCATGGCGGCGCGCGTCTTTGCAGCGATGTCACGCGCCCGTATTTCCGTGGT\n",
            "GCTGATTACGCAATCATCTTCCGAATACAGCATCAGTTTCTGCGTTCCACAAAGCGACTGTGTGCGAGCTGAACGGGCAA\n",
            "TGCAGGAAGAGTTCTACCTGGAACTGAAAGAAGGCTTACTGGAGCCGCTGGCAGTGACGGAACGGCTGGCCATTATCTCG\n",
            "GTGGTAGGTGATGGTATGCGCACCTTGCGTGGGATCTCGGCGAAATTCTTTGCCGCACTGGCCCGCGCCAATATCAACAT\n",
            "TGTCGCCATTGCTCAGGGATCTTCTGAACGCTCAATCTCTGTCGTGGTAAATAACGATGATGCGACCACTGGCGTGCGCG\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wc -l amp_res_1.fastq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "K8-ipRBmlP1H",
        "outputId": "a48d0ce5-2381-48ba-b782-d538f65a9df9"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "1823504 amp_res_1.fastq\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wc -l amp_res_2.fastq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "VAEbyKgnlWrw",
        "outputId": "22811a81-3014-49ed-f708-ee9837a8f431"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "1823504 amp_res_2.fastq\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "1823504/4"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "VJncJqwBlgF5",
        "outputId": "b874f64b-8da7-41a2-fd62-9760bcf46c31"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "455876.0"
            ]
          },
          "metadata": {},
          "execution_count": 9
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#!apt-get install fastqc"
      ],
      "metadata": {
        "id": "oC09SbaggMPC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#!fastqc -h"
      ],
      "metadata": {
        "id": "d0bH3cm7dnEb"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!./FastQC/fastqc amp_res_1.fastq\n",
        "!./FastQC/fastqc amp_res_2.fastq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "qBH305Zwh1B2",
        "outputId": "a1e6cef3-74b9-4453-da77-18c339c96b4c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Started analysis of amp_res_1.fastq\n",
            "Approx 5% complete for amp_res_1.fastq\n",
            "Approx 10% complete for amp_res_1.fastq\n",
            "Approx 15% complete for amp_res_1.fastq\n",
            "Approx 20% complete for amp_res_1.fastq\n",
            "Approx 25% complete for amp_res_1.fastq\n",
            "Approx 30% complete for amp_res_1.fastq\n",
            "Approx 35% complete for amp_res_1.fastq\n",
            "Approx 40% complete for amp_res_1.fastq\n",
            "Approx 45% complete for amp_res_1.fastq\n",
            "Approx 50% complete for amp_res_1.fastq\n",
            "Approx 55% complete for amp_res_1.fastq\n",
            "Approx 60% complete for amp_res_1.fastq\n",
            "Approx 65% complete for amp_res_1.fastq\n",
            "Approx 70% complete for amp_res_1.fastq\n",
            "Approx 75% complete for amp_res_1.fastq\n",
            "Approx 80% complete for amp_res_1.fastq\n",
            "Approx 85% complete for amp_res_1.fastq\n",
            "Approx 90% complete for amp_res_1.fastq\n",
            "Approx 95% complete for amp_res_1.fastq\n",
            "Analysis complete for amp_res_1.fastq\n",
            "Started analysis of amp_res_2.fastq\n",
            "Approx 5% complete for amp_res_2.fastq\n",
            "Approx 10% complete for amp_res_2.fastq\n",
            "Approx 15% complete for amp_res_2.fastq\n",
            "Approx 20% complete for amp_res_2.fastq\n",
            "Approx 25% complete for amp_res_2.fastq\n",
            "Approx 30% complete for amp_res_2.fastq\n",
            "Approx 35% complete for amp_res_2.fastq\n",
            "Approx 40% complete for amp_res_2.fastq\n",
            "Approx 45% complete for amp_res_2.fastq\n",
            "Approx 50% complete for amp_res_2.fastq\n",
            "Approx 55% complete for amp_res_2.fastq\n",
            "Approx 60% complete for amp_res_2.fastq\n",
            "Approx 65% complete for amp_res_2.fastq\n",
            "Approx 70% complete for amp_res_2.fastq\n",
            "Approx 75% complete for amp_res_2.fastq\n",
            "Approx 80% complete for amp_res_2.fastq\n",
            "Approx 85% complete for amp_res_2.fastq\n",
            "Approx 90% complete for amp_res_2.fastq\n",
            "Approx 95% complete for amp_res_2.fastq\n",
            "Analysis complete for amp_res_2.fastq\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Число ридов в итоговом файле fastqc равно числу ридов, подсчитанному выше (455876). \\\\\n",
        "Мы можем видеть, что качество ридов значительно ухудшается начиная с ≈ 80 позиции. Среднее качество прочтения среди всех ридов равно ≈ 36; GC-состав в среднем равен 33%. Кроме того можно отметить, что качество прочтения первых позиций ридов достаточно низкое, а также что на концах ряда ридов детектируется адаптер Nextera transposase sequence. \\\\\n",
        "\n",
        "* Для того, чтобы решить проблему плохого качества прочтения в самом начале обычно для секвенирования берут фрагменты, которые будут покрывать интересующий нас регион, захватывая небольшие последовательности перед и после него, чтобы на центральный регион приходилось больше центральных частей ридов. \\\\\n",
        "* Проблема низкого качества ближе к концу ридов достаточно стандартная и решается путем уменьшения длины ридов и увеличением глубины прочтения.  \\\\\n",
        "* Часто на концах остаются адапторы (что мы и наблюдаем на графике), поэтому мы можем без потери качества подрезать концы ридов, и убрать знакомые нам последовательности адапторов.\n"
      ],
      "metadata": {
        "id": "ess59935bK3Z"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Тримминг"
      ],
      "metadata": {
        "id": "g8jcfQ14mA41"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get install trimmomatic"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Ttc1cXK2l-XR",
        "outputId": "f87f6dac-b732-41fa-89a0-a0873967fa3d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "The following package was automatically installed and is no longer required:\n",
            "  libnvidia-common-460\n",
            "Use 'apt autoremove' to remove it.\n",
            "The following additional packages will be installed:\n",
            "  libjbzip2-java\n",
            "The following NEW packages will be installed:\n",
            "  libjbzip2-java trimmomatic\n",
            "0 upgraded, 2 newly installed, 0 to remove and 4 not upgraded.\n",
            "Need to get 791 kB of archives.\n",
            "After this operation, 859 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libjbzip2-java all 0.9.1-4 [42.6 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu bionic/universe amd64 trimmomatic all 0.36+dfsg-3 [749 kB]\n",
            "Fetched 791 kB in 1s (627 kB/s)\n",
            "Selecting previously unselected package libjbzip2-java.\n",
            "(Reading database ... 125405 files and directories currently installed.)\n",
            "Preparing to unpack .../libjbzip2-java_0.9.1-4_all.deb ...\n",
            "Unpacking libjbzip2-java (0.9.1-4) ...\n",
            "Selecting previously unselected package trimmomatic.\n",
            "Preparing to unpack .../trimmomatic_0.36+dfsg-3_all.deb ...\n",
            "Unpacking trimmomatic (0.36+dfsg-3) ...\n",
            "Setting up libjbzip2-java (0.9.1-4) ...\n",
            "Setting up trimmomatic (0.36+dfsg-3) ...\n",
            "Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Запускаем **pair-ended** run"
      ],
      "metadata": {
        "id": "3Jv5S02FmKes"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!trimmomatic PE -phred33 amp_res_1.fastq amp_res_2.fastq output_1_paired.fastq output_1_unpaired.fastq output_2_paired.fastq output_2_unpaired.fastqc ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:2:True \\\n",
        "LEADING:20 TRAILING:20 SLIDINGWINDOW:10:20 MINLEN:20  "
      ],
      "metadata": {
        "id": "uXD0DjyAn8DD",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "8cb29ab1-ab7d-4ca2-cc28-b897b5aed671"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "TrimmomaticPE: Started with arguments:\n",
            " -phred33 amp_res_1.fastq amp_res_2.fastq output_1_paired.fastq output_1_unpaired.fastq output_2_paired.fastq output_2_unpaired.fastqc ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:2:True LEADING:20 TRAILING:20 SLIDINGWINDOW:10:20 MINLEN:20\n",
            "Multiple cores found: Using 2 threads\n",
            "java.io.FileNotFoundException: /content/TruSeq3-PE.fa (No such file or directory)\n",
            "\tat java.io.FileInputStream.open0(Native Method)\n",
            "\tat java.io.FileInputStream.open(FileInputStream.java:195)\n",
            "\tat java.io.FileInputStream.<init>(FileInputStream.java:138)\n",
            "\tat org.usadellab.trimmomatic.fasta.FastaParser.parse(FastaParser.java:54)\n",
            "\tat org.usadellab.trimmomatic.trim.IlluminaClippingTrimmer.loadSequences(IlluminaClippingTrimmer.java:110)\n",
            "\tat org.usadellab.trimmomatic.trim.IlluminaClippingTrimmer.makeIlluminaClippingTrimmer(IlluminaClippingTrimmer.java:71)\n",
            "\tat org.usadellab.trimmomatic.trim.TrimmerFactory.makeTrimmer(TrimmerFactory.java:32)\n",
            "\tat org.usadellab.trimmomatic.Trimmomatic.createTrimmers(Trimmomatic.java:59)\n",
            "\tat org.usadellab.trimmomatic.TrimmomaticPE.run(TrimmomaticPE.java:552)\n",
            "\tat org.usadellab.trimmomatic.Trimmomatic.main(Trimmomatic.java:80)\n",
            "Input Read Pairs: 455876 Both Surviving: 446259 (97.89%) Forward Only Surviving: 9216 (2.02%) Reverse Only Surviving: 273 (0.06%) Dropped: 128 (0.03%)\n",
            "TrimmomaticPE: Completed successfully\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Input Read Pairs: 455876 Both Surviving: 446259 (97.89%) Forward Only Surviving: 9216 (2.02%) Reverse Only Surviving: 273 (0.06%) Dropped: 128 (0.03%) \\\\\n",
        "\n",
        "Эти числа совпадают с числами при применении команды *wc -l*: 1785036/4 = 446259"
      ],
      "metadata": {
        "id": "8pNkL8LGoU2_"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wc -l output_1_paired.fastq"
      ],
      "metadata": {
        "id": "5YQQJdQfqOUY",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "8dc1397c-97a2-4a24-a31d-74f323ad458e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "1785036 output_1_paired.fastq\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "1785036/4"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PgI6uavmqZVA",
        "outputId": "f3a3daf1-8a12-48c6-b77a-5d8550945f66"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "446259.0"
            ]
          },
          "metadata": {},
          "execution_count": 14
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#!fastqc -o . /content/amp_res_1_trimmed.fastq /content/amp_res_2_trimmed.fastq"
      ],
      "metadata": {
        "id": "u0bVQi4-Hn4U"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "PWBBF50gI7mI",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "bc5136fe-7ef9-45aa-f974-4324f5907d83"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Started analysis of output_1_paired.fastq\n",
            "Approx 5% complete for output_1_paired.fastq\n",
            "Approx 10% complete for output_1_paired.fastq\n",
            "Approx 15% complete for output_1_paired.fastq\n",
            "Approx 20% complete for output_1_paired.fastq\n",
            "Approx 25% complete for output_1_paired.fastq\n",
            "Approx 30% complete for output_1_paired.fastq\n",
            "Approx 35% complete for output_1_paired.fastq\n",
            "Approx 40% complete for output_1_paired.fastq\n",
            "Approx 45% complete for output_1_paired.fastq\n",
            "Approx 50% complete for output_1_paired.fastq\n",
            "Approx 55% complete for output_1_paired.fastq\n",
            "Approx 60% complete for output_1_paired.fastq\n",
            "Approx 65% complete for output_1_paired.fastq\n",
            "Approx 70% complete for output_1_paired.fastq\n",
            "Approx 75% complete for output_1_paired.fastq\n",
            "Approx 80% complete for output_1_paired.fastq\n",
            "Approx 85% complete for output_1_paired.fastq\n",
            "Approx 90% complete for output_1_paired.fastq\n",
            "Approx 95% complete for output_1_paired.fastq\n",
            "Analysis complete for output_1_paired.fastq\n",
            "Started analysis of output_2_paired.fastq\n",
            "Approx 5% complete for output_2_paired.fastq\n",
            "Approx 10% complete for output_2_paired.fastq\n",
            "Approx 15% complete for output_2_paired.fastq\n",
            "Approx 20% complete for output_2_paired.fastq\n",
            "Approx 25% complete for output_2_paired.fastq\n",
            "Approx 30% complete for output_2_paired.fastq\n",
            "Approx 35% complete for output_2_paired.fastq\n",
            "Approx 40% complete for output_2_paired.fastq\n",
            "Approx 45% complete for output_2_paired.fastq\n",
            "Approx 50% complete for output_2_paired.fastq\n",
            "Approx 55% complete for output_2_paired.fastq\n",
            "Approx 60% complete for output_2_paired.fastq\n",
            "Approx 65% complete for output_2_paired.fastq\n",
            "Approx 70% complete for output_2_paired.fastq\n",
            "Approx 75% complete for output_2_paired.fastq\n",
            "Approx 80% complete for output_2_paired.fastq\n",
            "Approx 85% complete for output_2_paired.fastq\n",
            "Approx 90% complete for output_2_paired.fastq\n",
            "Approx 95% complete for output_2_paired.fastq\n",
            "Analysis complete for output_2_paired.fastq\n"
          ]
        }
      ],
      "source": [
        "!./FastQC/fastqc output_1_paired.fastq\n",
        "!./FastQC/fastqc output_2_paired.fastq"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Мы можем видеть, что после тримминга параметр per base sequence quality заметно улучшился - вы видим, что квартили теперь уже и ближе к медиане и все находятся в \"зеленой зоне\" с качеством выше 28."
      ],
      "metadata": {
        "id": "pT1shWfiqN6T"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Картирование ридов, получение sam и bam файлов"
      ],
      "metadata": {
        "id": "mblNinNJsLIH"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get install bwa"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "oNMBjpB1qkUS",
        "outputId": "26dd8e79-2730-48c1-aaac-b999a2fa4b20"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "The following package was automatically installed and is no longer required:\n",
            "  libnvidia-common-460\n",
            "Use 'apt autoremove' to remove it.\n",
            "The following NEW packages will be installed:\n",
            "  bwa\n",
            "0 upgraded, 1 newly installed, 0 to remove and 4 not upgraded.\n",
            "Need to get 176 kB of archives.\n",
            "After this operation, 404 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 bwa amd64 0.7.17-1ubuntu0.1 [176 kB]\n",
            "Fetched 176 kB in 1s (336 kB/s)\n",
            "Selecting previously unselected package bwa.\n",
            "(Reading database ... 125440 files and directories currently installed.)\n",
            "Preparing to unpack .../bwa_0.7.17-1ubuntu0.1_amd64.deb ...\n",
            "Unpacking bwa (0.7.17-1ubuntu0.1) ...\n",
            "Setting up bwa (0.7.17-1ubuntu0.1) ...\n",
            "Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!bwa index"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "q5SB1H0usKDM",
        "outputId": "618e9ad9-2c56-4a48-bfcf-e6df9523c77d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Usage:   bwa index [options] <in.fasta>\n",
            "\n",
            "Options: -a STR    BWT construction algorithm: bwtsw, is or rb2 [auto]\n",
            "         -p STR    prefix of the index [same as fasta name]\n",
            "         -b INT    block size for the bwtsw algorithm (effective with -a bwtsw) [10000000]\n",
            "         -6        index files named as <in.fasta>.64.* instead of <in.fasta>.* \n",
            "\n",
            "Warning: `-a bwtsw' does not work for short genomes, while `-a is' and\n",
            "         `-a div' do not work not for long genomes.\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!bwa index GCF_000005845.2_ASM584v2_genomic.fna"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "q9wS2f_tsVaf",
        "outputId": "90f63a4e-ffd1-47db-dfdd-2ce8dbb94521"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[bwa_index] Pack FASTA... 0.02 sec\n",
            "[bwa_index] Construct BWT for the packed sequence...\n",
            "[bwa_index] 0.70 seconds elapse.\n",
            "[bwa_index] Update BWT... 0.02 sec\n",
            "[bwa_index] Pack forward-only FASTA... 0.02 sec\n",
            "[bwa_index] Construct SA from BWT and Occ... 0.29 sec\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa index GCF_000005845.2_ASM584v2_cds_from_genomic.fna\n",
            "[main] Real time: 1.144 sec; CPU: 1.058 sec\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!bwa mem GCF_000005845.2_ASM584v2_genomic.fna output_1_paired.fastq output_2_paired.fastq > alignment.sam "
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "eUyhyjiW7Uxy",
        "outputId": "a3f1293d-db18-4343-ddb1-798cecc51c6a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
            "[M::process] read 106336 sequences (10000178 bp)...\n",
            "[M::process] read 108264 sequences (10000141 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6, 41216, 1, 11)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (137, 176, 221)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 389)\n",
            "[M::mem_pestat] mean and std.dev: (180.19, 63.30)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 473)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (82, 117, 185)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 391)\n",
            "[M::mem_pestat] mean and std.dev: (123.27, 60.10)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 494)\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 106336 reads in 3.155 CPU sec, 3.108 real sec\n",
            "[M::process] read 108866 sequences (10000007 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (15, 41709, 0, 10)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (71, 94, 125)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 233)\n",
            "[M::mem_pestat] mean and std.dev: (96.43, 48.02)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 289)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (135, 174, 219)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 387)\n",
            "[M::mem_pestat] mean and std.dev: (178.14, 62.81)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 471)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (61, 165, 186)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 436)\n",
            "[M::mem_pestat] mean and std.dev: (140.60, 79.03)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 561)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 108264 reads in 4.381 CPU sec, 4.399 real sec\n",
            "[M::process] read 106080 sequences (10000082 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (10, 42027, 0, 7)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (94, 144, 178)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 346)\n",
            "[M::mem_pestat] mean and std.dev: (144.50, 75.07)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 445)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (134, 173, 218)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 386)\n",
            "[M::mem_pestat] mean and std.dev: (177.25, 62.29)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 470)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_process_seqs] Processed 108866 reads in 3.673 CPU sec, 3.522 real sec\n",
            "[M::process] read 105898 sequences (10000056 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (9, 41257, 0, 8)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (136, 176, 222)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 394)\n",
            "[M::mem_pestat] mean and std.dev: (180.59, 63.98)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 480)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 106080 reads in 4.891 CPU sec, 5.107 real sec\n",
            "[M::process] read 106462 sequences (10000139 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (8, 41120, 1, 9)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (135, 174, 218)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 384)\n",
            "[M::mem_pestat] mean and std.dev: (178.18, 62.49)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 467)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 105898 reads in 4.014 CPU sec, 4.173 real sec\n",
            "[M::process] read 107332 sequences (10000102 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (12, 41260, 0, 12)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (80, 117, 175)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 365)\n",
            "[M::mem_pestat] mean and std.dev: (116.67, 58.99)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 460)\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (135, 175, 221)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 393)\n",
            "[M::mem_pestat] mean and std.dev: (179.72, 63.37)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 479)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (51, 101, 211)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 531)\n",
            "[M::mem_pestat] mean and std.dev: (112.33, 78.53)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 691)\n",
            "[M::mem_pestat] skip orientation FF\n",
            "[M::mem_pestat] skip orientation RR\n",
            "[M::mem_process_seqs] Processed 106462 reads in 3.513 CPU sec, 3.396 real sec\n",
            "[M::process] read 104428 sequences (10000060 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (8, 41304, 0, 7)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (135, 175, 220)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 390)\n",
            "[M::mem_pestat] mean and std.dev: (178.68, 62.84)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 475)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 107332 reads in 3.587 CPU sec, 3.372 real sec\n",
            "[M::process] read 38852 sequences (3652257 bp)...\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (9, 40611, 0, 7)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (135, 175, 220)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 390)\n",
            "[M::mem_pestat] mean and std.dev: (179.48, 63.34)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 475)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 104428 reads in 3.423 CPU sec, 3.261 real sec\n",
            "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (1, 14926, 1, 2)\n",
            "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
            "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
            "[M::mem_pestat] (25, 50, 75) percentile: (133, 171, 216)\n",
            "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 382)\n",
            "[M::mem_pestat] mean and std.dev: (175.75, 61.64)\n",
            "[M::mem_pestat] low and high boundaries for proper pairs: (1, 465)\n",
            "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
            "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
            "[M::mem_process_seqs] Processed 38852 reads in 1.237 CPU sec, 1.104 real sec\n",
            "[main] Version: 0.7.17-r1188\n",
            "[main] CMD: bwa mem GCF_000005845.2_ASM584v2_cds_from_genomic.fna output_1_paired.fastq output_2_paired.fastq\n",
            "[main] Real time: 32.034 sec; CPU: 31.979 sec\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools view -S -b alignment.sam > alignment.bam"
      ],
      "metadata": {
        "id": "Yt58sIKi4iqj"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools flagstat alignment.bam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fXF4AQKc5tOv",
        "outputId": "8de2de4d-2b41-4eb7-e794-25e27dc947f4"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "901331 + 0 in total (QC-passed reads + QC-failed reads)\n",
            "0 + 0 secondary\n",
            "8813 + 0 supplementary\n",
            "0 + 0 duplicates\n",
            "798247 + 0 mapped (88.56% : N/A)\n",
            "892518 + 0 paired in sequencing\n",
            "446259 + 0 read1\n",
            "446259 + 0 read2\n",
            "715618 + 0 properly paired (80.18% : N/A)\n",
            "759578 + 0 with itself and mate mapped\n",
            "29856 + 0 singletons (3.35% : N/A)\n",
            "42972 + 0 with mate mapped to a different chr\n",
            "41407 + 0 with mate mapped to a different chr (mapQ>=5)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Overall, the 82.02% (88.56%) of reads are mapped."
      ],
      "metadata": {
        "id": "_EqbEe39vUWH"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "*901331 + 0 in total (QC-passed reads + QC-failed reads)                                                                              0 + 0 secondary                                                                                                                      8813 + 0 supplementary                                                                                                               0 + 0 duplicates                                                                                                                     798247 + 0 mapped (88.56% : N/A)                                                                                                     892518 + 0 paired in sequencing                                                                                                      446259 + 0 read1                                                                                                                     446259 + 0 read2                                                                                                                     715618 + 0 properly paired (80.18% : N/A)                                                                                            759578 + 0 with itself and mate mapped                                                                                               29856 + 0 singletons (3.35% : N/A)                                                                                                   42972 + 0 with mate mapped to a different chr                                                                                        41407 + 0 with mate mapped to a different chr (mapQ>=5)*"
      ],
      "metadata": {
        "id": "Dy6JfzFtvEZR"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools sort alignment.bam -o alignment_sorted.bam"
      ],
      "metadata": {
        "id": "gbc7ny0I6GCS"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools index alignment_sorted.bam"
      ],
      "metadata": {
        "id": "-6yzhTJVwMyy"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Variant calling "
      ],
      "metadata": {
        "id": "VlCMuCjFW_NW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools mpileup -f GCF_000005845.2_ASM584v2_genomic.fna alignment_sorted.bam >  my.mpileup                            "
      ],
      "metadata": {
        "id": "xYYpFbeowV4p",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "06d62b16-d32e-48ff-fc61-90a13dc16bdb"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[mpileup] 1 samples in 1 input files\n",
            "<mpileup> Set max per-file depth to 8000\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Появился файл mpileup с сводкой по числу изменений по каждой позиции "
      ],
      "metadata": {
        "id": "1XUxdQFGYUro"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!curl -L -k -o VarScan.v2.4.2.jar https://github.com/dkoboldt/varscan/releases/download/2.4.2/VarScan.v2.4.2.jar\n",
        "!java -jar VarScan.v2.4.2.jar"
      ],
      "metadata": {
        "id": "QYM92KS0YSfU",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "4171b986-3979-44ae-f7ea-2143c78cb594"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "curl: /usr/local/lib/libcurl.so.4: no version information available (required by curl)\n",
            "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
            "                                 Dload  Upload   Total   Spent    Left  Speed\n",
            "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\n",
            "100  112k  100  112k    0     0   299k      0 --:--:-- --:--:-- --:--:--  299k\n",
            "VarScan v2.4.2\n",
            "\n",
            "***NON-COMMERCIAL VERSION***\n",
            "\n",
            "USAGE: java -jar VarScan.jar [COMMAND] [OPTIONS] \n",
            "\n",
            "COMMANDS:\n",
            "\tpileup2snp\t\tIdentify SNPs from a pileup file\n",
            "\tpileup2indel\t\tIdentify indels a pileup file\n",
            "\tpileup2cns\t\tCall consensus and variants from a pileup file\n",
            "\tmpileup2snp\t\tIdentify SNPs from an mpileup file\n",
            "\tmpileup2indel\t\tIdentify indels an mpileup file\n",
            "\tmpileup2cns\t\tCall consensus and variants from an mpileup file\n",
            "\n",
            "\tsomatic\t\t\tCall germline/somatic variants from tumor-normal pileups\n",
            "\tcopynumber\t\t\tDetermine relative tumor copy number from tumor-normal pileups\n",
            "\treadcounts\t\tObtain read counts for a list of variants from a pileup file\n",
            "\n",
            "\tfilter\t\t\tFilter SNPs by coverage, frequency, p-value, etc.\n",
            "\tsomaticFilter\t\tFilter somatic variants for clusters/indels\n",
            "\tfpfilter\t\tApply the false-positive filter\n",
            "\n",
            "\tprocessSomatic\t\tIsolate Germline/LOH/Somatic calls from output\n",
            "\tcopyCaller\t\tGC-adjust and process copy number changes from VarScan copynumber output\n",
            "\tcompare\t\t\tCompare two lists of positions/variants\n",
            "\tlimit\t\t\tRestrict pileup/snps/indels to ROI positions\n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!java -jar VarScan.v2.4.2.jar mpileup2snp -h"
      ],
      "metadata": {
        "id": "LzDXUpG9bTgw",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "1beb53ac-df30-4c93-e4da-b58ee6f1a1bf"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Only SNPs will be reported\n",
            "Warning: No p-value threshold provided, so p-values will not be calculated\n",
            "Min coverage:\t8\n",
            "Min reads2:\t2\n",
            "Min var freq:\t0.2\n",
            "Min avg qual:\t15\n",
            "P-value thresh:\t0.01\n",
            "USAGE: java -jar VarScan.jar mpileup2cns [pileup file] OPTIONS\n",
            "\tmpileup file - The SAMtools mpileup file\n",
            "\n",
            "\tOPTIONS:\n",
            "\t--min-coverage\tMinimum read depth at a position to make a call [8]\n",
            "\t--min-reads2\tMinimum supporting reads at a position to call variants [2]\n",
            "\t--min-avg-qual\tMinimum base quality at a position to count a read [15]\n",
            "\t--min-var-freq\tMinimum variant allele frequency threshold [0.01]\n",
            "\t--min-freq-for-hom\tMinimum frequency to call homozygote [0.75]\n",
            "\t--p-value\tDefault p-value threshold for calling variants [99e-02]\n",
            "\t--strand-filter\tIgnore variants with >90% support on one strand [1]\n",
            "\t--output-vcf\tIf set to 1, outputs in VCF format\n",
            "\t--vcf-sample-list\tFor VCF output, a list of sample names in order, one per line\n",
            "\t--variants\tReport only variant (SNP/indel) positions [0]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Only SNPs will be reported. Warning: No p-value threshold provided, so p-values will not be calculated. \n",
        "* Min coverage: 8\n",
        "* Min reads2: 2\n",
        "* Min var freq: 0.2\n",
        "* Min avg qual: 15\n",
        "* P-value thresh: 0.01 \n",
        "\n"
      ],
      "metadata": {
        "id": "fPrSSN3_d1tE"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Повысим требования к минимальному проценту оснований, отличающихся от референсных на каждой позиции, который требуется для идентификации мутации по этой позиции. По умолчанию это значение равно 0.01 (1%), возьмем 0.1 (10%)."
      ],
      "metadata": {
        "id": "RS279ba1fxT2"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!java -jar VarScan.v2.4.2.jar  mpileup2snp my.mpileup --min-var-freq 0.1 --variants --output-vcf 1 > VarScan_results_10.vcf"
      ],
      "metadata": {
        "id": "IatRCibLbXgz"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "* 9 variant positions (6 SNP, 3 indel)\n",
        "* 1 were failed by the strand-filter\n",
        "* 5 variant positions reported (5 SNP, 0 indel)\n",
        "\n",
        "Мы видим, что при таком пороге программа обнаружила 6 SNP. Посмотрим, сколько оснований преодолели порог в 50%:\n"
      ],
      "metadata": {
        "id": "9cG6JwH0f-l-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!java -jar VarScan.v2.4.2.jar  mpileup2snp my.mpileup --min-var-freq 0.5 --variants --output-vcf 1 > VarScan_results_50.vcf"
      ],
      "metadata": {
        "id": "eHcJZru0iuhZ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "* 9 variant positions (6 SNP, 3 indel)\n",
        "* 1 were failed by the strand-filter \n",
        "* 5 variant positions reported (5 SNP, 0 indel) \n",
        "\n",
        "Ничего не изменилось - нам остается только проаннотировать эти мутации!"
      ],
      "metadata": {
        "id": "ALyRT0NGjTcq"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!mkdir -p data/k12\n",
        "!cd ./data/k12"
      ],
      "metadata": {
        "id": "9SlzEqAR1XE1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.gbff.gz"
      ],
      "metadata": {
        "id": "d7OJ1otv1Lrj"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!gunzip GCF_000005845.2_ASM584v2_genomic.gbff.gz "
      ],
      "metadata": {
        "id": "xbgBIdIw1ZJC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!nano snpEff.config # write k12.genome : ecoli_K12 there"
      ],
      "metadata": {
        "id": "4oNldihK1Phn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!cp GCF_000005845.2_ASM584v2_genomic.gbff data/k12/genes.gbk \n",
        "!cd ../../"
      ],
      "metadata": {
        "id": "ZITgeaGK1bgC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!java -jar snpEff.jar build -genbank -v k12"
      ],
      "metadata": {
        "id": "GoI5NFuY1d9g"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!java -jar snpEff.jar ann k12 VarScan_results_50.vcf > VarScan_results_annotated.vcf "
      ],
      "metadata": {
        "id": "r2jZTVI0hyKa"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Genes\n"
      ],
      "metadata": {
        "id": "atTdW2oC0B5d"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Printing nice table with frequencies:"
      ],
      "metadata": {
        "id": "XNLZYDnjSx9I"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!cat VarScan_results_annotated.vcf"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-TOWUW7Dj0tr",
        "outputId": "44ba6429-4135-4f5f-a4cc-3073a87dd782"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "##fileformat=VCFv4.1\n",
            "##source=VarScan2\n",
            "##INFO=<ID=ADP,Number=1,Type=Integer,Description=\"Average per-sample depth of bases with Phred score >= 15\">\n",
            "##INFO=<ID=WT,Number=1,Type=Integer,Description=\"Number of samples called reference (wild-type)\">\n",
            "##INFO=<ID=HET,Number=1,Type=Integer,Description=\"Number of samples called heterozygous-variant\">\n",
            "##INFO=<ID=HOM,Number=1,Type=Integer,Description=\"Number of samples called homozygous-variant\">\n",
            "##INFO=<ID=NC,Number=1,Type=Integer,Description=\"Number of samples not called\">\n",
            "##FILTER=<ID=str10,Description=\"Less than 10% or more than 90% of variant supporting reads on one strand\">\n",
            "##FILTER=<ID=indelError,Description=\"Likely artifact due to indel reads at this position\">\n",
            "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
            "##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">\n",
            "##FORMAT=<ID=SDP,Number=1,Type=Integer,Description=\"Raw Read Depth as reported by SAMtools\">\n",
            "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Quality Read Depth of bases with Phred score >= 15\">\n",
            "##FORMAT=<ID=RD,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases (reads1)\">\n",
            "##FORMAT=<ID=AD,Number=1,Type=Integer,Description=\"Depth of variant-supporting bases (reads2)\">\n",
            "##FORMAT=<ID=FREQ,Number=1,Type=String,Description=\"Variant allele frequency\">\n",
            "##FORMAT=<ID=PVAL,Number=1,Type=String,Description=\"P-value from Fisher's Exact Test\">\n",
            "##FORMAT=<ID=RBQ,Number=1,Type=Integer,Description=\"Average quality of reference-supporting bases (qual1)\">\n",
            "##FORMAT=<ID=ABQ,Number=1,Type=Integer,Description=\"Average quality of variant-supporting bases (qual2)\">\n",
            "##FORMAT=<ID=RDF,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases on forward strand (reads1plus)\">\n",
            "##FORMAT=<ID=RDR,Number=1,Type=Integer,Description=\"Depth of reference-supporting bases on reverse strand (reads1minus)\">\n",
            "##FORMAT=<ID=ADF,Number=1,Type=Integer,Description=\"Depth of variant-supporting bases on forward strand (reads2plus)\">\n",
            "##FORMAT=<ID=ADR,Number=1,Type=Integer,Description=\"Depth of variant-supporting bases on reverse strand (reads2minus)\">\n",
            "##SnpEffVersion=\"5.1d (build 2022-04-19 15:49), by Pablo Cingolani\"\n",
            "##SnpEffCmd=\"SnpEff  k12 VarScan_results_50.vcf \"\n",
            "##INFO=<ID=ANN,Number=.,Type=String,Description=\"Functional annotations: 'Allele | Annotation | Annotation_Impact | Gene_Name | Gene_ID | Feature_Type | Feature_ID | Transcript_BioType | Rank | HGVS.c | HGVS.p | cDNA.pos / cDNA.length | CDS.pos / CDS.length | AA.pos / AA.length | Distance | ERRORS / WARNINGS / INFO' \">\n",
            "##INFO=<ID=LOF,Number=.,Type=String,Description=\"Predicted loss of function effects for this variant. Format: 'Gene_Name | Gene_ID | Number_of_transcripts_in_gene | Percent_of_transcripts_affected'\">\n",
            "##INFO=<ID=NMD,Number=.,Type=String,Description=\"Predicted nonsense mediated decay effects for this variant. Format: 'Gene_Name | Gene_ID | Number_of_transcripts_in_gene | Percent_of_transcripts_affected'\">\n",
            "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tSample1\n",
            "NC_000913.3\t93043\t.\tC\tG\t.\tPASS\tADP=17;WT=0;HET=0;HOM=1;NC=0;ANN=G|missense_variant|MODERATE|ftsI|b0084|transcript|b0084|protein_coding|1/1|c.1631C>G|p.Ala544Gly|1631/1767|1631/1767|544/588||,G|upstream_gene_variant|MODIFIER|murE|b0085|transcript|b0085|protein_coding||c.-123C>G|||||123|WARNING_TRANSCRIPT_NO_START_CODON,G|upstream_gene_variant|MODIFIER|murF|b0086|transcript|b0086|protein_coding||c.-1607C>G|||||1607|,G|upstream_gene_variant|MODIFIER|mraY|b0087|transcript|b0087|protein_coding||c.-2959C>G|||||2959|,G|upstream_gene_variant|MODIFIER|murD|b0088|transcript|b0088|protein_coding||c.-4044C>G|||||4044|,G|downstream_gene_variant|MODIFIER|cra|b0080|transcript|b0080|protein_coding||c.*4011C>G|||||4011|WARNING_TRANSCRIPT_NO_START_CODON,G|downstream_gene_variant|MODIFIER|mraZ|b0081|transcript|b0081|protein_coding||c.*2951C>G|||||2951|,G|downstream_gene_variant|MODIFIER|rsmH|b0082|transcript|b0082|protein_coding||c.*2008C>G|||||2008|,G|downstream_gene_variant|MODIFIER|ftsL|b0083|transcript|b0083|protein_coding||c.*1646C>G|||||1646|\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:93:18:17:0:17:100%:4.2852E-10:0:36:0:0:6:11\n",
            "NC_000913.3\t482698\t.\tT\tA\t.\tPASS\tADP=16;WT=0;HET=0;HOM=1;NC=0;ANN=A|missense_variant|MODERATE|acrB|b0462|transcript|b0462|protein_coding|1/1|c.1706A>T|p.Gln569Leu|1706/3150|1706/3150|569/1049||,A|upstream_gene_variant|MODIFIER|pdeB|b0457|transcript|b0457|protein_coding||c.-4081A>T|||||4081|WARNING_TRANSCRIPT_NO_START_CODON,A|upstream_gene_variant|MODIFIER|ylaC|b0458|transcript|b0458|protein_coding||c.-3447A>T|||||3447|,A|upstream_gene_variant|MODIFIER|maa|b0459|transcript|b0459|protein_coding||c.-2780A>T|||||2780|,A|upstream_gene_variant|MODIFIER|hha|b0460|transcript|b0460|protein_coding||c.-2390A>T|||||2390|,A|upstream_gene_variant|MODIFIER|tomB|b0461|transcript|b0461|protein_coding||c.-1990A>T|||||1990|,A|upstream_gene_variant|MODIFIER|acrR|b0464|transcript|b0464|protein_coding||c.-3063T>A|||||3063|,A|upstream_gene_variant|MODIFIER|mscK|b0465|transcript|b0465|protein_coding||c.-3838T>A|||||3838|,A|downstream_gene_variant|MODIFIER|acrA|b0463|transcript|b0463|protein_coding||c.*1728A>T|||||1728|\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:87:16:16:0:16:100%:1.6637E-9:0:45:0:0:12:4\n",
            "NC_000913.3\t852762\t.\tA\tG\t.\tPASS\tADP=14;WT=0;HET=0;HOM=1;NC=0;ANN=G|upstream_gene_variant|MODIFIER|glnH|b0811|transcript|b0811|protein_coding||c.-4758T>C|||||4758|,G|upstream_gene_variant|MODIFIER|dps|b0812|transcript|b0812|protein_coding||c.-3851T>C|||||3851|,G|upstream_gene_variant|MODIFIER|rhtA|b0813|transcript|b0813|protein_coding||c.-2665T>C|||||2665|,G|upstream_gene_variant|MODIFIER|opgE|b0815|transcript|b0815|protein_coding||c.-165T>C|||||165|,G|upstream_gene_variant|MODIFIER|mntR|b0817|transcript|b0817|protein_coding||c.-421A>G|||||421|,G|upstream_gene_variant|MODIFIER|ybiR|b0818|transcript|b0818|protein_coding||c.-885A>G|||||885|,G|upstream_gene_variant|MODIFIER|ybiT|b0820|transcript|b0820|protein_coding||c.-3201A>G|||||3201|WARNING_TRANSCRIPT_NO_START_CODON,G|downstream_gene_variant|MODIFIER|yliM|b4736|transcript|b4736|protein_coding||c.*2365A>G|||||2365|,G|downstream_gene_variant|MODIFIER|ompX|b0814|transcript|b0814|protein_coding||c.*1797A>G|||||1797|,G|downstream_gene_variant|MODIFIER|mntS|b4705|transcript|b4705|protein_coding||c.*107T>C|||||107|,G|downstream_gene_variant|MODIFIER|ldtB|b0819|transcript|b0819|protein_coding||c.*2062T>C|||||2062|,G|intragenic_variant|MODIFIER|rybA|b4416|gene_variant|b4416|||n.852762A>G||||||\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:76:14:14:0:14:100%:2.4927E-8:0:36:0:0:9:5\n",
            "NC_000913.3\t3535147\t.\tA\tC\t.\tPASS\tADP=18;WT=0;HET=0;HOM=1;NC=0;ANN=C|missense_variant|MODERATE|envZ|b3404|transcript|b3404|protein_coding|1/1|c.722T>G|p.Val241Gly|722/1353|722/1353|241/450||,C|upstream_gene_variant|MODIFIER|yhgE|b3402|transcript|b3402|protein_coding||c.-2708T>G|||||2708|,C|upstream_gene_variant|MODIFIER|greB|b3406|transcript|b3406|protein_coding||c.-1665A>C|||||1665|,C|upstream_gene_variant|MODIFIER|yhgF|b3407|transcript|b3407|protein_coding||c.-2238A>C|||||2238|,C|downstream_gene_variant|MODIFIER|hslO|b3401|transcript|b3401|protein_coding||c.*4495A>C|||||4495|,C|downstream_gene_variant|MODIFIER|pck|b3403|transcript|b3403|protein_coding||c.*707A>C|||||707|,C|downstream_gene_variant|MODIFIER|ompR|b3405|transcript|b3405|protein_coding||c.*718T>G|||||718|\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:99:18:18:0:18:100%:1.1019E-10:0:34:0:0:12:6\n",
            "NC_000913.3\t4390754\t.\tG\tT\t.\tPASS\tADP=15;WT=0;HET=0;HOM=1;NC=0;ANN=T|synonymous_variant|LOW|rsgA|b4161|transcript|b4161|protein_coding|1/1|c.756C>A|p.Ala252Ala|756/1053|756/1053|252/350||WARNING_TRANSCRIPT_NO_START_CODON,T|upstream_gene_variant|MODIFIER|mscM|b4159|transcript|b4159|protein_coding||c.-1384C>A|||||1384|WARNING_TRANSCRIPT_NO_START_CODON,T|upstream_gene_variant|MODIFIER|psd|b4160|transcript|b4160|protein_coding||c.-394C>A|||||394|WARNING_TRANSCRIPT_NO_START_CODON,T|upstream_gene_variant|MODIFIER|orn|b4162|transcript|b4162|protein_coding||c.-850G>T|||||850|,T|upstream_gene_variant|MODIFIER|yjeV|b4670|transcript|b4670|protein_coding||c.-2138G>T|||||2138|,T|upstream_gene_variant|MODIFIER|nnr|b4167|transcript|b4167|protein_coding||c.-3312G>T|||||3312|,T|upstream_gene_variant|MODIFIER|tsaE|b4168|transcript|b4168|protein_coding||c.-4831G>T|||||4831|,T|downstream_gene_variant|MODIFIER|yjeO|b4158|transcript|b4158|protein_coding||c.*4736G>T|||||4736|,T|downstream_gene_variant|MODIFIER|queG|b4166|transcript|b4166|protein_coding||c.*2174C>A|||||2174|\tGT:GQ:SDP:DP:RD:AD:FREQ:PVAL:RBQ:ABQ:RDF:RDR:ADF:ADR\t1/1:81:16:15:0:15:100%:6.4467E-9:0:36:0:0:8:7\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cat VarScan_results_annotated.vcf | awk 'NR>29 {print $1, $2, $4, $5, $8, $10}' > genes_vcf.txt"
      ],
      "metadata": {
        "id": "MbeoV6y232C-"
      },
      "execution_count": 19,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "\n",
        "df = pd.read_csv(\"genes_vcf.txt\", sep=\" \", header=None, names=[\"Gene\", \"Position\", \"Before\", \"After\", \"Info\", \"Freq\"])"
      ],
      "metadata": {
        "id": "VXcXZErDkEeJ"
      },
      "execution_count": 20,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "QontCZpUkG4k",
        "outputId": "763c5800-7bf5-4d8b-9635-cd0cbcdefb1e"
      },
      "execution_count": 21,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "          Gene  Position Before After  \\\n",
              "0  NC_000913.3     93043      C     G   \n",
              "1  NC_000913.3    482698      T     A   \n",
              "2  NC_000913.3    852762      A     G   \n",
              "3  NC_000913.3   3535147      A     C   \n",
              "4  NC_000913.3   4390754      G     T   \n",
              "\n",
              "                                                Info  \\\n",
              "0  ADP=17;WT=0;HET=0;HOM=1;NC=0;ANN=G|missense_va...   \n",
              "1  ADP=16;WT=0;HET=0;HOM=1;NC=0;ANN=A|missense_va...   \n",
              "2  ADP=14;WT=0;HET=0;HOM=1;NC=0;ANN=G|upstream_ge...   \n",
              "3  ADP=18;WT=0;HET=0;HOM=1;NC=0;ANN=C|missense_va...   \n",
              "4  ADP=15;WT=0;HET=0;HOM=1;NC=0;ANN=T|synonymous_...   \n",
              "\n",
              "                                              Freq  \n",
              "0  1/1:93:18:17:0:17:100%:4.2852E-10:0:36:0:0:6:11  \n",
              "1   1/1:87:16:16:0:16:100%:1.6637E-9:0:45:0:0:12:4  \n",
              "2    1/1:76:14:14:0:14:100%:2.4927E-8:0:36:0:0:9:5  \n",
              "3  1/1:99:18:18:0:18:100%:1.1019E-10:0:34:0:0:12:6  \n",
              "4    1/1:81:16:15:0:15:100%:6.4467E-9:0:36:0:0:8:7  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-25d4a2d5-e3fc-4dcc-9a53-10032a5ba843\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Gene</th>\n",
              "      <th>Position</th>\n",
              "      <th>Before</th>\n",
              "      <th>After</th>\n",
              "      <th>Info</th>\n",
              "      <th>Freq</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>NC_000913.3</td>\n",
              "      <td>93043</td>\n",
              "      <td>C</td>\n",
              "      <td>G</td>\n",
              "      <td>ADP=17;WT=0;HET=0;HOM=1;NC=0;ANN=G|missense_va...</td>\n",
              "      <td>1/1:93:18:17:0:17:100%:4.2852E-10:0:36:0:0:6:11</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>NC_000913.3</td>\n",
              "      <td>482698</td>\n",
              "      <td>T</td>\n",
              "      <td>A</td>\n",
              "      <td>ADP=16;WT=0;HET=0;HOM=1;NC=0;ANN=A|missense_va...</td>\n",
              "      <td>1/1:87:16:16:0:16:100%:1.6637E-9:0:45:0:0:12:4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>NC_000913.3</td>\n",
              "      <td>852762</td>\n",
              "      <td>A</td>\n",
              "      <td>G</td>\n",
              "      <td>ADP=14;WT=0;HET=0;HOM=1;NC=0;ANN=G|upstream_ge...</td>\n",
              "      <td>1/1:76:14:14:0:14:100%:2.4927E-8:0:36:0:0:9:5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>NC_000913.3</td>\n",
              "      <td>3535147</td>\n",
              "      <td>A</td>\n",
              "      <td>C</td>\n",
              "      <td>ADP=18;WT=0;HET=0;HOM=1;NC=0;ANN=C|missense_va...</td>\n",
              "      <td>1/1:99:18:18:0:18:100%:1.1019E-10:0:34:0:0:12:6</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>NC_000913.3</td>\n",
              "      <td>4390754</td>\n",
              "      <td>G</td>\n",
              "      <td>T</td>\n",
              "      <td>ADP=15;WT=0;HET=0;HOM=1;NC=0;ANN=T|synonymous_...</td>\n",
              "      <td>1/1:81:16:15:0:15:100%:6.4467E-9:0:36:0:0:8:7</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-25d4a2d5-e3fc-4dcc-9a53-10032a5ba843')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-25d4a2d5-e3fc-4dcc-9a53-10032a5ba843 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-25d4a2d5-e3fc-4dcc-9a53-10032a5ba843');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 21
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "df[\"Gene\"] = df[\"Info\"].apply(lambda x: x.split(\";\")[5][4:].split(\"|\")[3])\n",
        "df[\"Effect\"] = df[\"Info\"].apply(lambda x: x.split(\";\")[5][4:].split(\"|\")[1])\n",
        "df[\"Impact\"] = df[\"Info\"].apply(lambda x: x.split(\";\")[5][4:].split(\"|\")[2])\n",
        "df[\"GeneID\"] = df[\"Info\"].apply(lambda x: x.split(\";\")[5][4:].split(\"|\")[4])\n",
        "df[\"Freq\"] = df[\"Freq\"] = df[\"Freq\"].apply(lambda x: float(x.split(\":\")[6][:-1]))\n",
        "df.dtypes"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "t9OVhpVUPTB_",
        "outputId": "7f6d2a75-243e-4547-f845-337c8908e9c9"
      },
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Gene         object\n",
              "Position      int64\n",
              "Before       object\n",
              "After        object\n",
              "Info         object\n",
              "Freq        float64\n",
              "Effect       object\n",
              "Impact       object\n",
              "GeneID       object\n",
              "dtype: object"
            ]
          },
          "metadata": {},
          "execution_count": 23
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "df.drop(columns = [\"Info\"], inplace = True)"
      ],
      "metadata": {
        "id": "uijrqYx2ndTa"
      },
      "execution_count": 25,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 214
        },
        "id": "nOTu5Qc3QDbL",
        "outputId": "b1686d8c-ba04-4d37-e064-cd2efc00a01f"
      },
      "execution_count": 26,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "   Gene  Position Before After   Freq                 Effect    Impact GeneID\n",
              "0  ftsI     93043      C     G  100.0       missense_variant  MODERATE  b0084\n",
              "1  acrB    482698      T     A  100.0       missense_variant  MODERATE  b0462\n",
              "2  glnH    852762      A     G  100.0  upstream_gene_variant  MODIFIER  b0811\n",
              "3  envZ   3535147      A     C  100.0       missense_variant  MODERATE  b3404\n",
              "4  rsgA   4390754      G     T  100.0     synonymous_variant       LOW  b4161"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-4b21585e-5775-49d8-9116-f6aa2ba018cb\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Gene</th>\n",
              "      <th>Position</th>\n",
              "      <th>Before</th>\n",
              "      <th>After</th>\n",
              "      <th>Freq</th>\n",
              "      <th>Effect</th>\n",
              "      <th>Impact</th>\n",
              "      <th>GeneID</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>ftsI</td>\n",
              "      <td>93043</td>\n",
              "      <td>C</td>\n",
              "      <td>G</td>\n",
              "      <td>100.0</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>b0084</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>acrB</td>\n",
              "      <td>482698</td>\n",
              "      <td>T</td>\n",
              "      <td>A</td>\n",
              "      <td>100.0</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>b0462</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>glnH</td>\n",
              "      <td>852762</td>\n",
              "      <td>A</td>\n",
              "      <td>G</td>\n",
              "      <td>100.0</td>\n",
              "      <td>upstream_gene_variant</td>\n",
              "      <td>MODIFIER</td>\n",
              "      <td>b0811</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>envZ</td>\n",
              "      <td>3535147</td>\n",
              "      <td>A</td>\n",
              "      <td>C</td>\n",
              "      <td>100.0</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>b3404</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>rsgA</td>\n",
              "      <td>4390754</td>\n",
              "      <td>G</td>\n",
              "      <td>T</td>\n",
              "      <td>100.0</td>\n",
              "      <td>synonymous_variant</td>\n",
              "      <td>LOW</td>\n",
              "      <td>b4161</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-4b21585e-5775-49d8-9116-f6aa2ba018cb')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-4b21585e-5775-49d8-9116-f6aa2ba018cb button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-4b21585e-5775-49d8-9116-f6aa2ba018cb');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 26
        }
      ]
    }
  ]
}
